Arbitrary self types v2 (#3519)
rust-lang/rfcs: text/3519-arbitrary-self-types-v2.md

- Feature Name: Arbitrary Self Types 2.0
- Start Date: 2023-05-04
- RFC PR: [rust-lang/rfcs#3519](https://github.com/rust-lang/rfcs/pull/3519)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

Allow types that implement the new `trait Receiver<Target=Self>` to be the receiver of a method.

# Motivation
[motivation]: #motivation

Today, methods can only be received by value, by reference, or by one of a few blessed smart pointer types from `core`, `alloc` and `std` (`Arc<Self>`, `Box<Self>`, `Pin<P>` and `Rc<Self>`).

It's been assumed that this will eventually be generalized to support any smart pointer, such as an `CustomPtr<Self>`. Since late 2017, it has been available on nightly under the `arbitrary_self_types` feature for types that implement `Deref<Target=Self>` and for raw pointers.

This RFC proposes some changes to the existing nightly feature based on the experience gained, with a view towards stabilizing the feature in the relatively near future.

## Motivation for the arbitrary self types feature overall

> One use-case is cross-language interop (JavaScript, Python, C++), where other languages' references can’t guarantee the aliasing and exclusivity semantics required of a Rust reference. For example, the C++ `this` pointer can't be practically or safely represented as a Rust reference because C++ may retain other pointers to the data and it might mutate at any time. Yet, calling C++ methods intrinsically requires a `this` reference.

Reviewer: This needs expanding, since it sounds like arbitrary self types are mainly useful for unsafe code interacting with FFI? That would be very niche.

I think what you are hinting at here is that such code could define a newtype around a raw pointer and then provide a completely safe API? If so, then please don't just hint at this -- actually spell out an example, showing the desired Rust API.

Author: Added - thanks - you're right that we're aiming to use this from safe Rust code.

Another case is when the existence of a reference is, itself, semantically important — for example, reference counting, or if relayout of a UI should occur each time a mutable reference ceases to exist. In these cases it's not OK to allow a regular Rust reference to exist, and yet sometimes we still want to be able to call methods on a reference-like thing.

> In theory, users can define their own smart pointers. In practice, they're second-class citizens compared to the smart pointers in Rust's standard library. User-defined smart pointers to `T` can accept method calls only if the receiver (`self`) type is `&T` or `&mut T`, which isn't acceptable if we can't safely create native Rust references to a `T`.

Reviewer 2: There's also another angle to this which I feel is important to recognize, which is that these smart pointers *can be stacked* with ordinary references, providing information both about the underlying type and the smart pointer.

For example, taking `&Arc<T>` allows me to both clone the smart pointer (noting that the underlying `T` might not implement `Clone`) in addition to access the data inside the type, which is useful for some methods. Also, being able to change a method from accepting `&self` to `self: &Arc<Self>` can be done in a mostly frictionless way, whereas changing from `&self` to a static method accepting `&Arc<Self>` will always require some amount of refactoring.

The "unable to create `&T`" angle is certainly something we want to support, but it's also worth mentioning that many users have reasons to want to borrow or even move their smart pointer types with a `self` parameter even if they could create a reference to the underlying value.

Author: I agree. Thanks, I added an extra paragraph.

This RFC proposes to loosen this restriction to allow custom smart pointer types to be accepted as a `self` type just like for the standard library types.

See also [this blog post](https://medium.com/@adetaylor/the-case-for-stabilizing-arbitrary-self-types-b07bab22bb45), especially for a list of more specific use-cases.

## Motivation for the v2 changes

Unstable Rust contains an implementation of arbitrary self types based around the `Deref<Target=T>` trait. Naturally, that trait also provides a means to create a `&T`.

However, if it's OK to create a reference `&T`, you _probably don't need this feature_. You can probably simply use `&self` as your receiver type.

This feature is fundamentally aimed at smart pointer types `P<T>` where it's not safe to create a reference `&T`. As noted above, that's most commonly because of semantic differences to pointers in other languages, but it might be because references have special meaning or behavior in some pure Rust domain. Either way, it's not OK to create a Rust reference `&T` or `&mut T`, yet we may want to allow methods to be called on some reference-like thing.

> For this reason, implementing `Deref::deref` is problematic for _nearly everyone who wants to use arbitrary self types_.

Reviewer: This looks like unnecessary hyperbole.  `Rc<Self>` is clearly useful, despite needing `Deref`.

Author: Yeah, @clarfonthey 's comment above has opened my eyes to the usefulness of this feature even in cases where it's OK to create `&T`. I've reworded.

If you're implementing a smart pointer `P<T>` yet you can't allow a reference `&T` to exist, any option for implementing `Deref::deref` has drawbacks:

* Specify `Deref::Target=T` and panic in `Deref::deref`. Not good.

> * Specify `Deref::Target=*const T`. This works with the current arbitrary self types feature, but is only possible if your smart pointer type contains a `*const T` which you can reference - this isn't the case for (for instance) weak pointers or types containing `NonNull`.

Reviewer: Earlier the RFC says arbitrary-self needs `Target=Self`, but now `*const Self` also seems to be an option? This is inconsistent and I don't know which part of the RFC is right.

Author: Fixed. The earlier wording was indeed imprecise. I also added an example of how the current arbitrary self types is used.

Therefore, the current Arbitrary Self Types v2 provides a separate `Receiver` trait.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

When declaring a method, users can declare the type of the `self` receiver to be any type `T` where `T: Receiver<Target = Self>` or `Self`.

The `Receiver` trait is simple and only requires to specify the `Target` type:

```rust
trait Receiver {
    type Target: ?Sized;
}
```

The `Receiver` trait is already implemented for many standard library types:
- smart pointers in the standard library: `Rc<Self>`, `Arc<Self>`, `Box<Self>`, and `Pin<Ptr<Self>>` (and in fact, any type which implements `Deref`)
- references: `&Self` and `&mut Self`
- pointers: `*const Self` and `*mut Self`

Shorthand exists for references, so that `self` with no ascription is of type `Self`, `&self` is of type `&Self` and `&mut self` is of type `&mut Self`.

All of the following self types are valid:

```rust
impl Foo {
    fn by_value(self /* self: Self */);
    fn by_ref(&self /* self: &Self */);
    fn by_ref_mut(&mut self /* self: &mut Self */);

>     fn by_ptr(*const Self);

Reviewer 3: Is this meant to be `self: *const Self`, or a new shorthand for pointer self params `*const self`?

Author: Fixed, thanks!

    fn by_mut_ptr(self: *mut Self);
    fn by_box(self: Box<Self>);
    fn by_rc(self: Rc<Self>);
    fn by_custom_ptr(self: CustomPtr<Self>);
}

struct CustomPtr<T>(*const T);

impl<T> Receiver for CustomPtr<T> {
    type Target = T;
}
```

## Recursive arbitrary receivers

> Receivers are recursive and therefore allowed to be nested. If type `T` implements `Receiver<Target=U>`, and type `U` implements `Receiver<Target=Self>`, `T` is a valid receiver (and so on outward).

Reviewer: What is the motivation for this?

Also explain why is it done "outside" the trait -- otherwise we get overlapping instances.

Author: Hm. Good question. This recursiveness was a feature of the existing unstable `arbitrary_self_types` feature. The same overlappingness applies to `Deref` in just the same way as `Receiver` - we're not changing anything here.

For the inter-language interop use-cases, we don't have an immediate _need_ for the recursiveness, but my sense is that for other use cases it might be useful. The example below of `Rc<Box<Self>>` might be a good one, or `Pin<Box<Self>>`.

@clarfonthey I think you might be able to explain this better than me? :)

Reviewer 4: This is already how the blessed subset works, so keeping this is required to get rid of the special-ness of the blessed std types 
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d12e5ac493a2b492da901c6b84c592c3
```rs
struct S;
impl S {
    fn f(self: std::pin::Pin<std::pin::Pin<&Self>>) {}
    fn f2(self: Box<Box<Self>>) {}
}
```

Reviewer: Okay, fair. Would be good to explain that this is not new but just describing existing behavior.

Reviewer 2: I put an example in another comment here describing how it works for `Deref`, using the example of a `Box<String>`: https://github.com/rust-lang/rfcs/pull/3519#discussion_r1380653042

Basically, because of how deref coercions work, the compiler can iteratively apply the deref (or in this case, mark a receiver), but a simple trait bound will only capture one iteration. This is definitely weird, but it's the status quo at least.

Author: Added a note that this is existing behavior.

For example, this self type is valid:

```rust
impl MyType {
     fn by_rc_to_box(self: Rc<Box<Self>>) { ... }
}
```

> 

Reviewer 2: It's worth commenting on how this might mean that *using* the `Receiver` trait is unreliable if you're not the compiler, since `Receiver<Target: Receiver<Target = Self>>` generally means that `Receiver<Target != Self>` (using pseudo-syntax).

I can't think of a circumstance where someone would *want* to bound their code by `Receiver`, but if they are, they're probably doing something wrong.

Reviewer 5: A similar restriction applies to `Deref`, so at least it's nothing new

Reviewer 2: Right, but still worth reiterating IMHO.

Reviewer 6: > A similar restriction applies to `Deref`, so at least it's nothing new

Is this restriction for Deref documented somewhere?

Author: > I can't think of a circumstance where someone would want to bound their code by `Receiver`, but if they are, they're probably doing something wrong.

I think this is in fact sort of what I'm hinting at in one of the suggestions about [diagnostics, below](https://github.com/rust-lang/rfcs/pull/3519/files#diff-085d4b194a1af0c50cdb08e0ec395782aab67202ffc9d626ca6574c480bcbc42R257).

Perhaps we could explicitly issue a warning for anyone who bounds anything by `Receiver`? WDYT?

Reviewer 5: `Deref` has the same issue yet `Deref` bounds sometimes make sense, e.g. on [`Pin::new`](https://doc.rust-lang.org/std/pin/struct.Pin.html#impl-Pin%3CP%3E).

Reviewer 2: > > A similar restriction applies to `Deref`, so at least it's nothing new
> 
> Is this restriction for Deref documented somewhere?

As far as I'm aware, not explicitly. But it's pretty simple to understand; think of the implementation for `Box<T>`:

```rust
impl<T> Deref for Box<T> {
    type Target = T;
    // ...
}
```

By this definition, we know that the `Target` for `Box<Box<T>>` is `Box<T>`. But the compiler is able to unwrap this definition to just target `T`, since it can apply `Deref` iteratively. This means that if, for example, you had the bound `Deref<Target = T>`, you could not pass `Box<Box<T>>`, since the target for this is `Box<T>`, not `T`.

I don't believe it's possible to specify a bound that would iteratively unwrap the target like this without creating a helper trait, and even then, it's not immediately clear how you'd write such a trait. Maybe there's a crate that does it that I don't know about.

Author: Thanks for this discussion. I've tried to summarize it in the RFC in 544f1afa.

Author: I'm going to mark this as Resolved but please do reopen if the above commit doesn't reflect things here. Thanks again!

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## `core` libs changes

The `Receiver` trait is made public (removing its `#[doc(hidden)])` attribute), exposing it under `core::ops`. It gains a `Target` associated type.

This trait marks types that can be used as receivers other than the `Self` type of an impl or trait definition.

```rust
pub trait Receiver {
    type Target: ?Sized;
}
```

A blanket implementation is provided for any type that implements `Deref`:

```rust
impl<P: ?Sized, T: ?Sized> Receiver for P
where
    P: Deref<Target = T>,
{
    type Target = T;
}

> ```

Reviewer 4: There is actually a problem here that I didn't realize back when we talked about this. Implementing `Receiver` is effectively a promise to not add any new methods to the implementing type (given the reasons already outlined here, as adding methods could break downstream users for these kinds of types). This impl means, this promise now propagates to `Deref` implementations as well which seems like a very much unwanted side effect, especially given all the current `Deref` impls out there.

Reviewer 7: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0ea4bedd4fc2cf5d3f0a633fbade68ac

Reviewer 4: No because you can't add inherent impls to a foreign type, so you can't add new methods to them (ignoring fundamental types). With the `Receiver` trait you can kind of do this though.
```rs
//- crate: foo
pub struct Foo<T>(pub T);
impl<T> Receiver for Foo<T> {
    type Target = T;
}
//- crate: bar (depends on foo)
use foo::Foo;
struct Bar;
impl Bar {
    fn foobar(self: Foo<Bar>) {}
}
fn main() {
    Foo(Bar).foobar();
}
```
Adding a `fn foobar(self)` method to `Foo` would break the crate `bar` in this case, so adding a method to a `Receiver` implementing type is a breaking change. And since `Deref` here implies implementing `Receiver`, the same issue arises from just `Deref` implementations, that is the same issue would occur if instead of the `Receiver` impl in that example we had
```rs
impl<T> Deref for Foo<T> {
    type Target = T;
    fn deref(&self) -> &Self::Target { &self.0 }
}
```

Reviewer 7: Discarding `Receiver` from this RFC for the moment, I'm still missing what makes you take the stance that `Deref` implementors on today's Rust don't already carry the same promise not to add new methods. As far as I understand, this problem already exists with `Deref`, so adding `Receiver` to all `Deref` types changes nothing.

Reviewer 4: Ah sorry I didn't really answer your question. Taking your example here and modifying it to be split across two crates (everything in one crate does not exhibit any semver problems for obvious reasons):
```rs

//- crate b
struct B;

impl B {
    fn foo(&self, x: i32) -> i32 {
        x
    }
}

//- crate: a depends on crate b
struct A(B);

impl std::ops::Deref for A {
    type Target = B;
    fn deref(&self) -> &B { 
        &self.0
    }
}

impl A {
    // As of Rust 1.73, this will shadow B::foo for A
    fn foo(&self) -> i32 {
        0
    }
}

fn main() {
    let a = A(B);
    a.foo(); // will always resolve to `A::foo` no matter the methods on `B` due to how method resolution works
}
```
Adding or removing the `foo` method has no effect here on the method resolution of objects of type `A`, as we will always discover the `A::foo` method first and then stop resolution as is. No crate that depends on these crates can later on "inject" new resolutions into this, unlike how the `Receiver` proposal here works

Author: Results from trying to reproduce this multi-crate method shadowing problem with the existing unstable `arbitrary_self_types` support:

```rust
// crate Foo
use core::ops::Deref;

pub struct Foo<T>(pub T);
impl<T> Deref for Foo<T> {
    type Target = T;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
impl<T> Foo<T> {
    pub fn foobar(self) {
        println!("Calling foo's foobar");
    }
}

// crate bar
use foo::Foo;
struct Bar;
impl Bar {
    fn foobar(self: Foo<Bar>) {println!("Calling bar's foobar");}
}
fn main() {
    Foo(Bar).foobar();
}
```

Results:
```
error[E0034]: multiple applicable items in scope
 --> src/main.rs:8:14
  |
8 |     Foo(Bar).foobar();
  |              ^^^^^^ multiple `foobar` found
  |
note: candidate #1 is defined in an impl for the type `Bar`
 --> src/main.rs:5:5
  |
5 |     fn foobar(self: Foo<Bar>) {println!("Calling bar's foobar");}
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^
  = note: candidate #2 is defined in an impl for the type `Foo<T>`
```

I think this does show that adding new methods to the `Foo` crate can break the `Bar` crate - even with the existing unstable `arbitrary_self_types`.

It seems to me that we can possibly solve this by amending the method resolution such that in ambiguous situations it always picks the one it would have picked prior to `arbitrary_self_types`, instead of reporting an ambiguity? (I'm not 100% following this thread though I must admit)

Reviewer 5: > No crate that depends on these crates can later on "inject" new resolutions into this, unlike how the Receiver proposal here works

I'm not sure I'm following you fully, but I don't think there is a problem here. A user crate can indeed add something that looks like a new method on e.g. `Box<A>`, but because the crate that defines `Box` cannot know about `A`, there is nowhere where it could call a function that might be resolved as the downstream method.

Reviewer 4: ```rs
//- crate: foo
pub struct Foo<T>(pub T);
impl<T> Receiver for Foo<T> {
    type Target = T;
}
```

```rs
//- crate: bar (depends on foo)
use foo::Foo;
pub struct Bar;
impl Bar {
    pub fn foobar(self: Foo<Bar>) {}
}
```
```rs
//- crate: qux (depends on foo)
fn main() {
    Foo(Bar).foobar(); // resolves to `Bar::foobar`
}
```

Now, if crate `foo` adds the following:
```rs
impl<T> Foo<T> {
    pub fn foobar(self: Foo<T>) {}
}
```
crate `qux` will fail to compile due to an ambiguity, so adding new methods to a type that implements `Receiver` is always a breaking change. As `Foo(Bar).foobar()` could now be referring to either `Bar::foobar` or `<Foo::<Bar>>::foobar`

Since `Deref` implies implementing `Receiver` with the proposal here, that means the same restrictions fall onto any type implementing `Deref`. That's what I'm trying to get at here. That is in your `Box<A>` example, it doesn't need to to know about the `A`, because we can still reach this via generics.

Reviewer 6: I guess the way out of this problem is to make `Receiver` opt-in, which means no blanket implementation for `Deref`. Then we need another mechanism to ensure that the targets of Receiver and Deref not diverge, but that probably can be achieved with ✨ compiler magic ✨.

Reviewer 8: I think adding an inherent method `foo` to `T` where `T: Receiver<Target = U>` is a *minor* breaking change since anyone trying to call `U::foo` with `v.foo()` can just write `U::foo(v)` instead. (iirc this has precedence in deeming new trait methods as minor breaking changes even though they can become ambiguous and need `Trait::method(v)` instead of `v.method()`)

Therefore, I see no reason to remove the blanket `T: Receiver` implementation for `T: Deref`.

Reviewer 9: I also agree that it's a breaking but non-major change. Adding a blanket impl for an existing trait, implementing a new method on an existing type, and newly implementing a trait for an existing type are all breaking changes (and ones deemed non-major) regardless of arbitrary self types.

Reviewer 5: Oh I see. Doesn't `Deref` already have this exact problem though? That's exactly why `Box`/`Rc`/etc have no normal methods, since adding a `.foo()` would break any `Foo` that has `fn foo(&self)`

Reviewer 4: I don't see how `Deref` already has this problem. `Box`, `Rc`, `Pin` etc have htis problem because they are already allowed as arbitrary receivers as they have been special cased. `Deref` has no say in that. With the RFC as written, all `Deref` implementations *will* have this problem. Anyways, even if this is considered a minor breakage only (which I'd argue against personally, it very much feels like a major breakage to me), the RFC should definitely talk about this, both in that implementing `Receiver` has this semver problem as well as that the blanket impl will extend it to `Deref` impls.

Reviewer 5: No I mean plain Deref. Let me spell it out:

```rust
//- crate: foo (depends on nothing)
pub struct Foo<T>(pub T);
impl<T> Deref for Foo<T> {
    type Target = T;
    fn deref(...) {...}
}
```
```rust
//- crate: bar (depends on nothing)
pub struct Bar;
impl Bar {
    pub fn foobar(&self) {}
}
```
```rust
//- crate: qux (depends on bar and foo)
fn main() {
    Foo(Bar).foobar(); // resolves to `Bar::foobar`
}
```
Now, if crate foo adds the following:
```rust
impl<T> Foo<T> {
    pub fn foobar(&self) {}
}
```
then `Foo(Bar).foobar()` in crate qux now resolves to `Foo<T>::foobar`. I imagine this is a major breaking change since now qux could still compile yet be using a completely different function.

This is the same problem; `Deref` pointers already have to deal with it today. Hence I conclude: adding Receiver into the mix should not change anything, all is good.

Reviewer 10: I agree with @Nadrieril, this is a non-issue for this RFC, since the problem is already present on `Deref` types.

The RFC actually [has a section that talks about this](https://github.com/adetaylor/rfcs/blob/arbitrary-self-types-v2/text/3519-arbitrary-self-types-v2.md#method-shadowing), I've updated that in https://github.com/rust-lang/rfcs/pull/3519#pullrequestreview-1733257247 to include a few more references to external resources that further explain the problem.

I'll also link to https://github.com/rust-lang/rust-clippy/issues/11820, which proposes a new clippy lint to catch inherent methods on smart pointers, to hopefully help implementers of such types (including in the future, implementers of `Receiver`) to avoid this pitfall.

(See [alternatives](#no-blanket-implementation) for discussion of the tradeoffs here.)

It is also implemented for `&T`, `&mut T`, `*const T` and `*mut T`.

## Compiler changes

> The existing Rust [reference section for method calls describes the algorithm for assembling method call candidates](https://doc.rust-lang.org/reference/expressions/method-call-expr.html). This algorithm changes in one simple way: instead of dereferencing types (using the `Deref<Target=T>`) trait, we use the new `Receiver<Target=T>` trait to determine the next step.

Reviewer 5: I think this is subtler than it looks: if `x: &Arc<T>` and I call `x.foo()`, the current algorithm will (simplified):
- look for a `foo(&self)` method on `Arc`
- if not found, check if `Arc` implements `Deref`
- if yes, recursively look for a `foo(&self)` method on `<Arc<T> as Deref>::Target` (i.e. `T`)
- if not found, check if `T` implements `Deref`
- etc.

Your proposed algorithm would:
- look for a `foo(&self)` method on `Arc`
- if not found, check if `Arc` implements `Receiver`
- if yes, recursively (?) look for a `foo(&Arc<Self>)` method on `<Arc<T> as Receiver>::Target`
- if not found, what do we do next?

The algorithm needs at some point to look for a `foo(&self)` on `T`. I think this requires special-casing `Deref` still

Reviewer 5: It gets fun with nesting: say `x: &Arc<Box<T>>`. We need to check:
- `foo(&self)` on `Arc<Box<T>>`
- `foo(&Arc<Self>)` on `Box<T>`
- `foo(&Arc<Box<Self>>)` on `T`
- `foo(&self)` on `Box<T>`
- `foo(&Box<Self>)` on `T`
- `foo(&self)` on `T`

We'd need to decide on which should take priority, I think the order I listed them in is sensible. An algorithm for this would look like "try more and more layers of `Receiver` until we can't anymore, then take one layer of `Deref`, then start again.

Reviewer 5: Ah, I had missed [the section](https://github.com/adetaylor/rfcs/blob/arbitrary-self-types-v2/text/3519-arbitrary-self-types-v2.md#explore-both-receiver-and-deref-chains-while-identifying-method-candidates) where you talk about this, sorry. I still don't quite understand how there's a way to avoid `Deref` entirely, even with a blanket `impl<P: Deref> Receiver for P`

Author: We went backwards and forwards on how much detail to put in this section of the RFC! In the end we decided to just state the specific small change we're making, instead of trying to describe the current algorithm in lots of detail.

The current unstable `arbtrary_self_types` complains of an ambiguity in the [case of your example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6eeafbe7e26a7d02f870c14f9d2ebf71).

```
error[E0034]: multiple applicable items in scope
  --> src/main.rs:54:7
   |
54 |     x.foo();
   |       ^^^ multiple `foo` found
   |
note: candidate #1 is defined in an impl for the type `Arcish<Boxish<T>>`
  --> src/main.rs:24:5
   |
24 |     fn foo(&self) {
   |     ^^^^^^^^^^^^^
note: candidate #2 is defined in an impl for the type `Boxish<T>`
  --> src/main.rs:30:5
   |
30 |     fn foo(self: &Arcish<Self>) {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #3 is defined in an impl for the type `Concrete`
  --> src/main.rs:39:5
   |
39 |     fn foo(self: &Arcish<Boxish<Self>>) {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

That probably needs to change - a [comment thread above notes that this can cause inter-crate compatibility breakages](https://github.com/rust-lang/rfcs/pull/3519/files#r1379664429).

Buuuutt... I don't _think_ there's a need for `Deref` here. I think with a blanket implementation `impl<P: Deref> Receiver for P` the method candidate selection algorithm should never need to consider `Deref` explicitly. Can you explain more about why you think it might?

Reviewer 5: Hmm, I thought my detailed steps above made my point clear ><. Let me try again.

I seem to understand that there's a fundamental difference between what `Deref` and `Receiver` mean in this context. `Receiver` on `P<T>` means "look for methods like `foo(P<Self>)` on `T`. `Deref` on `P<T>` means "look for methods like `foo(&self)` on `T`. `Receiver` only ever allows you to look for methods that take `P<T>`, the `P` must be mentioned. Contrast with `Deref`, which removes one layer of pointer.

Assume the resolution algorithm only uses `Receiver`, and we have the blanket `Deref` impl. Assume I have `x: Arc<T>` and I call `x.foo()`. What we clearly do want is to call `foo(Arc<Self>)` if there is one. I believe this is what the `Receiver` impl means. Yet if there is no such method, we also want to call `foo(&self)` on `T` if there is one! And I don't see how `Receiver` on its own allows this.

Author: OK - I see what you mean, and I can see the part of the description which I need to clarify. The proposed algorithm won't use the `Deref` trait at all, but it _will_ retain the current logic which searches for method candidates not only on `T` but also `&T` and `&mut T`. I'll write something about that.

Author: I clarified a little - hopefully that helps! Please let me know if not.

Reviewer 5: Small comment: You marked the question resolved so I almost missed that you had replied to me. I find it a good convention to let whoever asked the question mark it as "resolved", unless it's a really obvious comment like a typo

## Object safety

Receivers are object safe if they implement the (unstable) `core::ops::DispatchFromDyn` trait.

As not all receivers might want to permit object safety or are unable to support it, object safety should remain being encoded in a different trait than the here proposed `Receiver` trait, likely `DispatchFromDyn`.

This RFC does not propose any changes to `DispatchFromDyn`. Since `DispatchFromDyn` is unstable at the moment, object-safe receivers might be delayed until `DispatchFromDyn` is stabilized. `Receiver` is not blocked on further `DispatchFromDyn` work, since non-object-safe receivers already cover a big chunk of the use-cases.

## Lifetime elision

As discussed in the [motivation](#motivation), this new facility is _most likely_ to be used in cases where a standard reference can't normally be used. But in other cases a smart pointer self type might wrap a standard Rust reference, and thus might be parameterized by a lifetime.

> Lifetime elision works in the expected fashion:

Reviewer: How does it work if the pointer has more than one lifetime?

Author: I can't spot any problems here - [here's a playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=867bae8b8d1e9e97680f02e972a59fd7). This is of course with the current unstable `arbitrary_self_types` but nothing in this area should change. If you can think I should clarify something in the RFC text please let me know.

Reviewer: What I meant is: how does lifetime elision work then? There's no "expected fashion" for this case since it's not currently possible. So when I write `fn foo<'a, 'b, 'c>(self: MyPtr<'a, 'b, Self>, other: &'c T) -> &U`, then which lifetime does the returned reference get? It's ambiguous. I think this should be rejected and require an explicit lifetime.

Reviewer 2: I would imagine that the lifetime elision rules would be augmented this way:

If the self type has exactly one lifetime, then the rules behave the same as `&self`. So, this:
```rust
fn method(self: P<'_, Self>, one: &T, two: &T) -> &T
```
Is sugar for:
```rust
fn method<'selflt, 'a, 'b>(self: P<'selflt, Self>, one: &'a T, two: &'b T) -> &'selflt T
```

If the return type has no lifetime, then the inference assumes that all elided lifetimes are distinct, even if `self` has multiple lifetimes. So, this:

```rust
fn method(self: P<'_, '_, Self>, one: &T, two: &T) -> Box<T>
```
Is sugar for:
```rust
fn method<'a, 'b, 'c, 'd>(self: P<'a, 'b, Self>, one: &'c T, two: &'d T) -> Box<T>
```

If the self type has multiple lifetimes, then it's flagged as ambiguous if the return type has any lifetimes. So, this is ambiguous:
```rust
fn method(self: P<'_, '_, Self>, one: &T, two: &T) -> &T
```

Reviewer: > If the self type has multiple lifetimes, then it's flagged as ambiguous if the return type has any lifetimes. So, this is ambiguous:

Yeah sounds good. The same should probably occur if `self` has no lifetime (like it already does for `self: Self`, I assume).

Author: I rewrote this section to include an example much like that playground link. Please take another look!

Reviewer: > However, I still believe the existing lifetime elision rules work OK with no changes. [Here's a playground using the existing unstable arbitrary_self_types support](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=afe43820163195ff26454e74f3c2881c) - if there's any ambiguity, the compiler insists that a lifetime is manually specified.

Those are not the "existing rules". The existing rules are what is documented in the reference or prior RFCs.

You are saying the compiler already implements the rules you are suggesting, which is great, but that doesn't mean those rules have been subject to any kind of scrutiny, and unless you can point at existing documentation detailing those rules, it is the job of this RFC to document them.

```rust
struct SmartPtr<'a, T: ?Sized>(&'a T);

impl<'a, T: ?Sized> Receiver for SmartPtr<'a, T> {
    type Target = T;
}

struct MyType;

impl MyType {
    fn m(self: SmartPtr<Self>) {}
    fn n(self: SmartPtr<'_, Self>) {}
    fn o<'a>(self: SmartPtr<'a, Self>) {}
}
```

## Diagnostics

The existing branches in the compiler for "arbitrary self types" already emit excellent diagnostics. We will largely re-use them, with the following improvements:

- In the case where a self type is invalid because it doesn't implement `Receiver`, the existing excellent error message will be updated.
- An easy mistake is to implement `Receiver` for `P<T>`, forgetting to specify `T: ?Sized`. `P<Self>` then only works as a `self` parameter in traits `where Self: Sized`, an unusual stipulation. It's not obvious that `Sized`ness is the problem here, so we will identify this case specifically and produce an error giving that hint.

> - There are certain types which feel like they "should" implement `Receiver` but do not: `Weak` and `NotNull`. If these are encountered as a self type, we should produce a specific diagnostic explaining that they do not implement `Receiver` and suggesting that they could be wrapped in a newtype wrapper if method calls are important. This will require `Weak` and `NonNull` be marked as lang items so that the compiler is aware of the special nature of these types. (The authors of this RFC feel that these extra lang-items _are_ worthwhile to produce these improved diagnostics - if the reader disagrees, please let us know.)

Reviewer 11: Isn't this exactly what the [diagnostic items](https://rustc-dev-guide.rust-lang.org/diagnostics/diagnostic-items.html) machinery is for?

Reviewer: Yeah, they don't need to be lang items for this.

Author: That's awesome news, thanks, I will rewrite this bit.

- Under some circumstances, the compiler identifies method candidates but then discovers that the self type doesn't match. This results currently in a simple "mismatched types" error; we can provide a more specific error message here. The only known case is where a method is generic over `Receiver`, and the caller explicitly specifies the wrong type:
    ```rust
    #![feature(receiver_trait)]

    use std::ops::Receiver;

    struct SmartPtr<'a, T: ?Sized>(&'a T);

    impl<'a, T: ?Sized> Receiver for SmartPtr<'a, T> {
        type Target = T;
    }

    struct Foo(u32);
    impl Foo {

>         fn a<R: Receiver<Target=Self>>(self: R) { }

Reviewer: Is that really something we want to allow? Is it ever useful?

Author: No, and it doesn't work. It's just that the current error message isn't very good, so this section is proposing we spot this case specifically and produce better diagnostics.

Reviewer: > No, and it doesn't work.

Ah good. :) Receiver types should be concrete IMO, at least until we have a clear usecase.

Author: To be specific, the case which doesn't work is the fourth line of the `main` function in this example.

The third line *does* work - [here's a playground using the existing unstable arbitrary self types support](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=d014901f33cec65187e3c2be85e7a04e).

So, if we wanted to forbid generic `Receiver`s, we'd have to _actively block it_. I'm fine with that if the consensus is that it's a good idea. I haven't heard of any use-cases for generic receivers. But I think we'll need to add a justification to this RFC - is it fair to say that we should block this just to reduce unknown unknowns, or will we need something a bit more crisp? (I'd note that this has been possible with the unstable `arbitrary_self_types` support for years)

Reviewer: > is it fair to say that we should block this just to reduce unknown unknowns,

Yeah that seems like a reasonable argument to me. There is a risk of subtle interactions with method resolutions and other parts of the language when we can't even know the receiver type of a method, and we don't know a usecase, so there's no reason to work out the details of what exactly happens in this case.

Author: OK, fad04aee amends this RFC to add a proposal that we ban generic receiver types. If that's OK with you I'll resolve this conversation tomorrow.

* We don't know of any use-cases which the `Receiver<Target=T>` and blanket implementation for `Deref` do not allow.

If some use-case presents itself where a type _must_ implement `Deref` but not `Receiver`; or a use-case presents itself where `Deref` and `Receiver` _must_ have different `Target`s then we will have to consider this more complex option.

## Generic parameter

Change the trait definition to have a generic parameter instead of an associated type. There might be permutations here which could allow a single smart pointer type to dispatch method calls to multiple possible receivers - but this would add complexity, no known use case exists, and it might cause worst-case O(n^2) performance on method lookup.

## Do not enable for pointers

It would be possible to respect the `Receiver` trait without allowing dispatch onto raw pointers - they are essentially independent changes to the candidate deduction algorithm.

We don't want to encourage the use of raw pointers, and would prefer rather that raw pointers are wrapped in a custom smart pointer that encodes and documents the invariants. So, there's an argument not to add the raw pointer support.

> However, the current unstable `arbitrary_self_types` feature provides support for raw pointer receivers, and with years of experience no major concerns have been spotted. We would prefer not to deviate from the existing proposal more than necessary. Moreover, we are led to believe that raw pointer receivers are quite important for the future of safe Rust, because stacked borrows makes it illegal to materialize references in many positions, and there are a lot of operations (like going from a raw pointer to a raw pointer to a field) where users don't need to or want to do that. We think the utility of including raw pointer receivers outweighs the risks of tempting people to over-use raw pointers.

Reviewer 5: Doesn't that mean that adding methods on `*const T`, like [`add()`](https://doc.rust-lang.org/std/primitive.pointer.html#method.addr), is a breaking change? How has this been dealt with so far?

Author: Good point. There are various Rust standard library types - `Box`, `Arc` etc. - where by policy no new methods are added because it would be a breaking change. Instead, associated functions are added.

If arbitrary self types is enabled (either the existing nightly unstable version, or our slightly tweaked version here) then that policy would need to be extended to the raw pointer types as well.

The Rust community therefore needs to decide which is higher priority:
* being able to add new methods to raw pointer types; or
* being able to receive method calls by raw pointer.

Unless someone can see some workaround or compromise?

@Manishearth hello, I think you were especially keen that arbitrary self types continues to support raw pointers. Do you have any views?

(Personally I think this might be a good argument to enable arbitrary self types _without_ raw pointers, and therefore to require people to write their own newtype wrappers if they want method dispatch).

Author: (there's a little bit more discussion of this in [this comment thread below](https://github.com/rust-lang/rfcs/pull/3519/files#r1379378763) which I'm going to close to avoid duplication)

Reviewer 5: I wanted to reply to [your comment](https://github.com/rust-lang/rfcs/pull/3519#discussion_r1380057062) about changing priority: I think this new problem is a much better problem. The original issue is "the pointer type `*const T` can't add any new methods ever", the new issue is "crate A can add `foo(*const Self)` methods, except that if they're from the set of known `*const T` methods, this is a major semver change". This is a bit surprising but highly manageable.

What worries me more is the high confusion cost of making method resolution more complicated.

Author: I'm still thinking this through but I wanted to note that we Rust doesn't current shadow things invisibly; instead it gives an error message if there's ambiguity.

* For arbitrary self types which are not raw pointers: we [show the ambiguity as an error](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=218cc52ca6d8c0b584c3fa06f77b234f#:~:text=Permalink%20to%20the%20playground).
* For arbitrary self types which are raw pointers, we [show similar errors if there are ambiguities, and we even warn if there are known upcoming future ambiguities](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=ea541e54f00fa15729af54f7d40668f5):
```
   = warning: once this associated item is added to the standard library, the ambiguity may cause an error or change in behavior!
   = note: for more information, see issue #48919 <https://github.com/rust-lang/rust/issues/48919>
   = help: call with fully qualified syntax `MyType::addr(...)` to keep using the current method
   = help: add `#![feature(strict_provenance)]` to the crate attributes to enable `ptr::const_ptr::<impl *const T>::addr`
   = note: `#[warn(unstable_name_collisions)]` on by default
```

This RFC should be more explicit somewhere about the existence of these errors and warnings, especially in the Method Shadowing section - I'll adjust.

Of course, this doesn't solve the issue here: if Rust adds a new `<raw pointer>::repaint()` method, that will cause downstream crates to cease building if they had:

```rust
struct Bedroom;
impl Bedroom {
  fn repaint(self: *const Bedroom) {}
}
```

@Nadrieril 's proposal, as I understand it, is that if method resolution does result in ambiguities in such a situation, where raw pointers are involved, we don't show an error for the ambiguity but instead automatically pick `Bedroom::repaint` and show a warning.

I do agree that seems to be manageable and I don't think it makes method resolution _cognitively_ more complex, because any such situations will result in a nice clear warning explaining the situation. It might make the _code_ significantly more complex - I'll look into it.

Author: OK, a concrete proposal here.

Method resolution comes up with a list of applicable `candidates`. These are currently determined by (among other things) following the chain of `Deref`s (this RFC will change this slightly to follow a chain of `Receiver`s). The number of such "autoderef" hops is recorded.

We then determine whether there's a single correct candidate, or if we should show an error.

Currently, there are two circumstances already when [the code](https://github.com/rust-lang/rust/blob/e7b22851a2c97980fd4e8235f3255af4fe0a12d5/compiler/rustc_hir_typeck/src/method/probe.rs#L1308) will favor one candidate over another:

* If some of the candidates are unstable, it will favor the others
* If there are several `impl`s of the same trait (this should not be user-visible)

If after this analysis there are _still_ several traits, it emits the errors [shown here](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=218cc52ca6d8c0b584c3fa06f77b234f#:~:text=Permalink%20to%20the%20playground).

The proposal is to add a third case where it favors one of the candidates over another, if there are multiple. Specifically:

Examine the autoderef hop count on each candidate.
If one candidate has a higher autoderef hop count than all the others, pick that one (and show a warning).
Otherwise, proceed onwards to show an error.

The intention is to pick the candidate on the "innermost" type in any smart pointer, disregarding methods on the surrounding smart pointer types (but showing a warning about the problem).

Author: I had a go at implementing this in [this commit](https://github.com/rust-lang/rust/commit/74834988f38382ad94337b5583cc249fe522cfb7) (albeit without the warning yet). I'm pleased to say it doesn't seem to break any of the existing tests, so hopefully there aren't too many gotchas with this plan. Let me know y'alls thoughts and if nobody objects I will update the RFC to say this (and possibly broach the subject of supporting `NotNull` and `Weak` per [this discussion](https://github.com/rust-lang/rfcs/pull/3519#issuecomment-1824448078)).

Reviewer 12: > The intention is to pick the candidate on the "innermost" type in any smart pointer,

Isn't this a breaking change because it's the opposite of the order that's already implemented and [specified in the reference](https://doc.rust-lang.org/reference/expressions/method-call-expr.html), and occasionally relied on for the “autoref/autoderef specialization” technique?

Author: I'm honestly not sure, but I think it's OK, because the words in the reference which we're chaniging are:

> If this results in multiple possible candidates, then it is an error

Reviewer 5: We mustn't change what type would be picked by Deref indeed. My guess is that the rule we want would apply only in case of an ambiguity that involves a Receiver type, and would favor the Receiver method. This shouldn't be a breaking change if there isn't a Receiver involved, or if the Receiver has no inherent methods (like Box or Rc). I'd need to write it down clearly to be fully convinced that works though

Reviewer 5: I had an example the other day: say `x: &Arc<Box<T>>` and I call `x.foo()`. Here are the methods that could be called, in my proposed order of priority ("most specific first"):
1. `foo(&Arc<Box<Self>>)` on `T`
1. `foo(&Arc<Self>)` on `Box<T>`
1. `foo(&Box<Self>)` on `T`
1. `foo(&self)` on `Arc<T>`
1. `foo(&self)` on `Box<T>`
1. `foo(&self)` on `T`
1. any trait methods

If we want "adding inherent methods to a Receiver type is not a breaking change", we need 4, 5 and 6 to be the lowest priority before trait methods. Question is: does this order break anything compared to what's allowed on stable today?

The current implementation of `arbitrary_self_types` sometimes errors on some of these combinations. The interesting cases for us is when 3 and 5 are available, or 1 and 4. This currently errors as ambiguous; we would instead pick 3 and 1 respectively.

Reviewer 5: Ok so I did a thorough experiment (try out my madness [here](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=f310d10f58f72b703b22df7c9ce685ba)). It appears the algorithm picks 4 over 3 in my list above, which could appear to break my proposal. The rest of the order is ok, so today's method resolution algorithm is compatible with this ordering:

a. `foo(&Arc<Box<Self>>)` on `T`
b. `foo(&Arc<Self>)` on `Box<T>`
c. `foo(&self)` on `Arc<T>`
d. `foo(&Box<Self>)` on `T`
e. `foo(&self)` on `Box<T>`
f. `foo(&self)` on `T`

However! All is still fine. Indeed, if `Arc` was not `Deref`, we would only consider the following cases:

a. `foo(&Arc<Box<Self>>)` on `T`
b. `foo(&Arc<Self>)` on `Box<T>`
c. `foo(&self)` on `Arc<T>`

Here adding an inherent method to `Arc` would not be a breaking change. The reason adding an inherent method to `Arc` is breaking is because `Arc: Deref`, and we already know that adding methods to `Deref` types is a breaking change.

In short: if we take the current implementation of `arbitrary_self_types` and simply accept more cases that are currently errors, we can get the order above and the property that "adding inherent methods to a Receiver type is not a breaking change".

Moreover! Given that the only stable `Receiver` types don't have inherent methods (I think, right?), I believe we can even change this order a bit without breakage. It will only break users of the unstable `arbitrary_self_types` feature in that one corner case. So we can pick my initial "most specific first" order too if we prefer.

I think that proves that we're good? Did I miss anything?

Reviewer 5: I grepped through std: the stable `Receiver`s are exactly `Box`, `Rc`, `Arc`, `Pin`, `&T`, and `&mut T`. None of these have inherent methods, except `Pin`, which has e.g. [`Pin::as_ref()`](https://doc.rust-lang.org/nightly/std/pin/struct.Pin.html#method.as_ref). So we can't change the order. But we can still do the first thing

Reviewer 13: > @Manishearth hello, I think you were especially keen that arbitrary self types continues to support raw pointers. Do you have any views?

Missed this. Yeah, I think it's important for the ergonomics of unsafe code. Wrapper types are an okay workaround, as are free functions.

Raw pointers are basically the native rust version of "I want to be able to hold a `CppPointer<T>` without breaking it", so it would be ideal for 

I think `#[fundamental]`'s behavior may be useful for dealing with raw pointer receivers: I don't think there's anything new here that doesn't apply to `&T` as well. I think it's fine for Rust to need new editions for new methods on `*const T`; and there are some ways of making that work even without editions.

## Provide compiler support for dereferencing pointers

This RFC proposes to implement `Receiver` for `*mut T` and `*const T` within the library. This is slightly different from the unstable arbitrary self types support, which instead hard-codes pointer support into the candidate deduction algorithm in the compiler (because obviously `Deref` can't be implemented for pointers.)

We prefer the option of specifying behavior in the library using the normal trait, though it's a compatibility break for users of Rust who don't adopt the `core` crate (including compiler tests).

## Implement for `Weak` and `NonNull`

> `Weak<T>` and `NonNull<T>` were not supported by the prior unstable arbitrary self types support, but they share the property that it may be desirable to implement method calls to `T` using them as self types. Unfortunately they also share the property that these types have many Rust methods using `self`, `&self` or `&mut self`. If we added to the set of Rust methods in future, we'd [shadow any such method calls](#method-shadowing). We can't implement `Receiver` for these types unless we come up with a policy that all subsequent additions to these types would instead be associated functions. That would make the future APIs for these types a confusing mismash of methods and associated functions, and the extra user complexity doesn't seem merited.

Reviewer 5: This applies to `*const T` too; it has [many methods](https://doc.rust-lang.org/std/primitive.pointer.html). Allowing methods that take `*const Self` seems pretty important to this RFC, maybe we can tweak the resolution algorithm instead? I'm not sure what's a good solution here.

Reviewer 5: One option would be that method resolution that goes through `Receiver` has higher priority than inherent methods on the pointer type. I pointed out in my other comment that `Deref` still needs a special-case in the resolution algorithm, so we can keep the current priority for these. This would be pretty confusing to use tho.

We could even special-case raw ptrs, `NonNull` and the like in method resolution :')

Reviewer 2: I think it would probably be ideal to have a dedicated diagnostic for shadowing built-in pointer methods like this, if we go the route of adding pointer receivers. People would expect `addr` to work normally, and overriding it yourself seems like a bad idea.

    // also calls into C++
}

fn main() {
    // Rust code manipulating C++ objects via C++-semantics references
    let cpp_obj_reference: CppRef<ConcreteCppType> = get_cpp_reference();
    // cpp_obj_reference does not obey Rust reference semantics. Other
    // "references" to the same data may exist in the Rust or C++ domain.
    // But it can effectively be used as an opaque token to pass safely
    // through Rust back into C++
    let some_value: u32 = cpp_obj_reference.get_int_field();
    let some_field = cpp_obj_reference.get_more_complex_field();
    cpp_obj_reference.equals(&get_cpp_reference());
}
```

(fuller example [here](https://github.com/google/autocxx/blob/main/src/reference_wrapper.rs#L117), with various [trait-based attempts](#not-do-it) to work around the lack of arbitrary self types.)

Another case is when the existence of a reference is, itself, semantically important — for example, reference counting, or if relayout of a UI should occur each time a mutable reference ceases to exist. In these cases it's not OK to allow a regular Rust reference to exist, and yet sometimes we still want to be able to call methods on a reference-like thing.

> A third motivation is that taking smart pointer types as `self` parameters can enable functions to act on the smart pointer type, not just the underlying data. For example, taking `&Arc<T>` allows the functions to both clone the smart pointer (noting that the underlying `T` might not implement `Clone`) in addition to access the data inside the type, which is useful for some methods. Also, being able to change a method from accepting `&self` to `self: &Arc<Self>` can be done in a mostly frictionless way, whereas changing from `&self` to a static method accepting `&Arc<Self>` will always require some amount of refactoring. These options are currently open only to Rust's built-in smart pointer types, not to custom smart pointer types.

Reviewer 14: ```suggestion
A third motivation is that taking smart pointer types as `self` parameters can enable functions to act on the smart pointer type, not just the underlying data. For example, taking `&Arc<T>` allows the functions to both clone the smart pointer (noting that the underlying `T` might not implement `Clone`) in addition to access the data inside the type, which is useful for some methods; this also makes it ergonomic in more cases to make `Arc<SomeType>` explicit rather than having `SomeType` contain an `Arc` internally and have `Arc`-like `clone` semantics. Also, being able to change a method from accepting `&self` to `self: &Arc<Self>` can be done in a mostly frictionless way, whereas changing from `&self` to a static method accepting `&Arc<Self>` will always require some amount of refactoring. These options are currently open only to Rust's built-in smart pointer types, not to custom smart pointer types.
```

Adding another potential benefit: I've seen code that would have been clearer if written using an explicit `Arc<SomeType>`, but which has to wrap the `Arc` internally instead because `Arc<SomeType>` doesn't allow calling methods that can operate on the `Arc` (e.g. because the method needs to produce a `Weak` reference).

Author: Yep that's come up in some of the discussion, thanks.

If you're implementing a smart pointer `P<T>`, and you need to allow `impl T { fn method(self: P<T>) { ... }}`, yet you can't allow a reference `&T` to exist, any option for implementing `Deref::deref` has drawbacks:

* Specify `Deref::Target=T` and panic in `Deref::deref`. Not good.
* Specify `Deref::Target=*const T`. This is only possible if your smart pointer type contains a `*const T` which you can reference - this isn't the case for (for instance) weak pointers or types containing `NonNull`.

Therefore, the current Arbitrary Self Types v2 provides a separate `Receiver` trait, so that there's no need to provide an awkward `Deref::deref` implementation.

In addition, this v2 proposes to block generic receivers, which are currently allowed by the v1 (unstable) arbitrary self types feature. See the [diagnostics section for reasoning](#diagnostics).

Aside from these differences, Arbitrary Self Types v2 is similar to the existing unstable `arbitrary_self_types` feature, including in its support for raw pointers as method receivers.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

> When declaring a method, users can declare the type of the `self` receiver to be any type `T` where `T: Receiver<Target = Self>` or `Self`.

Reviewer 14: ```suggestion
When declaring a method, users can also declare the type of the `self` receiver to be any type `T` where `T: Receiver<Target = Self>`, in addition to using `Self` by value or reference.
```

> The `Receiver` trait is simple and only requires to specify the `Target` type:

Reviewer 14: ```suggestion
The `Receiver` trait is simple and only requires specifying the `Target` type:
```

```rust
trait Receiver {
    type Target: ?Sized;
}
```

The `Receiver` trait is already implemented for many standard library types:

> - smart pointers in the standard library: `Rc<Self>`, `Arc<Self>`, `Box<Self>`, and `Pin<Ptr<Self>>` (and in fact, any type which implements `Deref`)

Reviewer 14: ```suggestion
- smart pointers in the standard library: `Rc<Self>`, `Arc<Self>`, `Box<Self>`, and `Pin<SomeSmartPtr<Self>>` (and in fact, any type which implements `Deref`)
```

For example, this self type is valid:

```rust
impl MyType {
     fn by_rc_to_box(self: Rc<Box<Self>>) { ... }
}
```

The Rust language doesn't provide a way for user code to use this recursive property in generics or iteration, so this trait is unlikely to be useful except to the compiler. Nevertheless, we don't intend to _prevent_ use of the `Receiver` trait by user code: since the same recursive property applies to `Deref` yet it's been occasionally useful to [introduce `Deref` bounds](https://doc.rust-lang.org/std/pin/struct.Pin.html#method.new_unchecked).

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## `core` libs changes

The `Receiver` trait is made public (removing its `#[doc(hidden)])` attribute), exposing it under `core::ops`. It gains a `Target` associated type.

This trait marks types that can be used as receivers other than the `Self` type of an impl or trait definition.

```rust
pub trait Receiver {
    type Target: ?Sized;
}
```

A blanket implementation is provided for any type that implements `Deref`:

```rust
impl<P: ?Sized> Receiver for P
where
    P: Deref,
{
    type Target = <P as Deref>::Target;
}

> ```

Reviewer 15: Does the new `Receiver` trait also work when the target is embedded directly in the struct, rather than being an actual smart pointer? E.g.:
```Rust
struct Wrapper<T> {
    extra_data: Foo,
    inner: T,
}

impl MyStruct {
    fn method1(self: Wrapper<Self>);
    fn method2(self: Arc<Wrapper<Self>>);
}
```

Reviewer 5: Yep. Receiver doesn't care about the internals, you can implement it on whatever you want

## Compiler changes

The existing Rust [reference section for method calls describes the algorithm for assembling method call candidates](https://doc.rust-lang.org/reference/expressions/method-call-expr.html). This algorithm changes in one simple way: instead of dereferencing types (using the `Deref<Target=T>`) trait, we use the new `Receiver<Target=T>` trait to determine the next step.

(Note that the existing algorithm isn't quite as simple as following the chain of `Deref<Target=T>`. In particular, `&T` and `&mut T` are considered as candidates too at each step; this RFC does not change that.)

Because a blanket implementation is provided for users of the `Deref` trait and for `&T`/`&mut T`, the net behavior is similar. But this provides the opportunity for types which can't implement `Deref` to act as method receivers.

## Object safety

Receivers are object safe if they implement the (unstable) `core::ops::DispatchFromDyn` trait.

As not all receivers might want to permit object safety or are unable to support it, object safety should remain being encoded in a different trait than the here proposed `Receiver` trait, likely `DispatchFromDyn`.

> This RFC does not propose any changes to `DispatchFromDyn`. Since `DispatchFromDyn` is unstable at the moment, object-safe receivers might be delayed until `DispatchFromDyn` is stabilized. `Receiver` is not blocked on further `DispatchFromDyn` work, since non-object-safe receivers already cover a big chunk of the use-cases.

Reviewer 15: Rust for Linux also needs dyn upcasting, so we are using unstable `DispatchFromDyn` trait as well. But it sounds like that is out of scope for this RFC.

Reviewer 16: Dyn upcasting should be stabilized soon. It was scheduled for next release, but [had to be reverted due to a soundness bug](https://github.com/rust-lang/rust/issues/120222).

Reviewer 17: `DispatchFromDyn` and dyn upcasting are two different things. Dyn upcasting is `dyn TraitSub` => `dyn TraitSuper`. This was scheduled to be stabilized for the next release. `DispatchFromDyn` is what allows calling a method on a trait object. This one wasn't being stabilized, nor is it covered by this RFC. It's implementation is also rather fragile and not ready to allow arbitrary user implementations. (your type needs to have a very specific shape to not get an ICE when trying to use `DispatchFromDyn` on it)

Reviewer 15: @bjorn3 Thanks for clarifying. The one we need is `DispatchFromDyn`, and we are careful to only use it for structs of a specific shape.

- In the case where a self type is invalid because it doesn't implement `Receiver`, the existing excellent error message will be updated.
- An easy mistake is to implement `Receiver` for `P<T>`, forgetting to specify `T: ?Sized`. `P<Self>` then only works as a `self` parameter in traits `where Self: Sized`, an unusual stipulation. It's not obvious that `Sized`ness is the problem here, so we will identify this case specifically and produce an error giving that hint.
- There are certain types which feel like they "should" implement `Receiver` but do not: `Weak` and `NotNull`. If these are encountered as a self type, we should produce a specific diagnostic explaining that they do not implement `Receiver` and suggesting that they could be wrapped in a newtype wrapper if method calls are important. We hope this can be achieved with [diagnostic items](https://rustc-dev-guide.rust-lang.org/diagnostics/diagnostic-items.html).
- The current unstable arbitrary self types feature allows generic receivers. For instance,
  ```rust
  impl Foo {
    fn a<R: Deref<Target=Self>>(self: R) { }
  }
  ```

>   We don't know a use-case for this. There are several cases where this can result in misleading diagnostics. (For instance, if such a method is called with an incorrect type (for example `smart_ptr.a::<&Foo>()` instead of `smart_ptr.a::<Foo>()`). We could attempt to find and fix all those cases. However, we feel that generic receiver types might risk subtle interactions with method resolutions and other parts of the language. We think it is a safer choice to generate an error on any declaration of a generic `self` type.

Reviewer 14: > We don't know a use-case for this.

Can you please clarify whether you can have a concrete type that takes a generic parameter in addition to `Self`? For instance, can you have `self: Box<T, impl Allocator>`? That seems like it should work fine, as long as the top-level type is concrete.

If possible, I think we should allow that case, and just disallow the case where the top-level receiver type itself is not concrete.

Author: I'm starting to think that there are [use cases for fully generic receivers](https://medium.com/@adetaylor/can-cppref-t-be-ergonomic-c9cb1365bda1). Still thinking on it, but we might amend this bit more radically.

Reviewer 5: That feels controversial/nonobvious enough that I'd suggest keeping it for a later RFC. The RFC without it is already very useful

>   We don't know a use-case for this. There are several cases where this can result in misleading diagnostics. (For instance, if such a method is called with an incorrect type (for example `smart_ptr.a::<&Foo>()` instead of `smart_ptr.a::<Foo>()`). We could attempt to find and fix all those cases. However, we feel that generic receiver types might risk subtle interactions with method resolutions and other parts of the language. We think it is a safer choice to generate an error on any declaration of a generic `self` type.

Reviewer 15: How broad is this? We have one where case where we want a smart pointer with two generic parameters. For example:
```rs
struct ListArc<T: ?Sized, const ID: u64> {
    ptr: NonNull<ArcInner<T>>,
}

impl MyStruct {
    fn method1(self: ListArc<Self, 17>) {}
    fn method2<const ID: u64>(self: ListArc<Self, ID>) {}
}
```
Would `method1` be allowed? Would `method2` be allowed? What if `ID` was a type rather than a const generic?

Reviewer 5: That's fine because the concrete type is known

    type Pointee = T;

    fn read(&self) -> *const Self::Pointee {
        todo!()
    }

    fn write(&mut self, _value: *const Self::Pointee) {
        todo!()
    }
}

// --------------------------------------------------------

struct SomeForeignLanguageType;

impl ConcreteForeignLanguageRef<SomeForeignLanguageType> {
    fn m(&self) {
        todo!()
    }
}

trait Tr {
    type RustType;

    fn tm(self)
    where
        Self: ForeignLanguageRef<Pointee = Self::RustType>;
}

impl Tr for ConcreteForeignLanguageRef<SomeForeignLanguageType> {
    type RustType = SomeForeignLanguageType;
    fn tm(self) {}
}

fn main() {
    let a = ConcreteForeignLanguageRef(SomeForeignLanguageType);
    a.m();
    a.tm();
}
```

This successfully allows method calls to `m()` and even `tm()` without a reference to a `SomeForeignLanguageType` ever existing. However, due to the orphan rule, this forces every crate to have its own equivalent of `ConcreteForeignLanguageRef`. This workaround has been used by some interop tools, but use across multiple crates requires many generic parameters (`impl ForeignLanguageRef<Pointee=SomeForeignLanguageType>`).

> ## Always use `unsafe` when interacting with other languages

Author: I'd be happy to drop it, because discussion of FFI unsafety sometimes leads to a fair bit of bikeshedding.

However I'm aiming to head off a kind of FAQ here: "Why do you need all this? Can't you just use regular references and use `unsafe` to ensure they're correct? Then we can all just use `Deref` and we don't need any of this `Receiver` stuff."

Superficially I think that's an appealing option. It only falls apart when you realize that humans also can't guarantee the semantics of C++ references match Rust, because they simply don't.

So, that's why I put it in. If you still think it does more harm than good, I'll zap it!

## Motivation for the arbitrary self types feature overall

One use-case is cross-language interop (JavaScript, Python, C++), where other languages' references can’t guarantee the aliasing and exclusivity semantics required of a Rust reference. For example, the C++ `this` pointer can't be practically or safely represented as a Rust reference because C++ may retain other pointers to the data and it might mutate at any time. Yet, calling C++ methods intrinsically requires a `this` reference. With "arbitrary self types", smart pointer types can be created which obey foreign-language semantics and can be used in safe Rust code:

```rust
#[repr(transparent)]
#[derive(Clone)]
/// A C++ reference. Obeys C++ reference semantics, not Rust reference semantics.
/// There is no exclusivity; the underlying data may mutate, etc.
pub struct CppRef<T: ?Sized> {
    ptr: *const T,
}

impl<T: ?Sized> Receiver for CppRef<T> {
    type Target = T;
}

// generated by bindings generator
struct ConcreteCppType {
    // ...
}

// all generated by bindings generator; mostly calls into C++
impl ConcreteCppType {

>     fn some_cpp_method(self: CppRef<Self>) {

Reviewer: So, the "obvious" alternative here is to not separately expose `CppRef` and `ConcreteCppType`, but instead expose `ConcreteCppTypeRef`, and write `self` methods on that. Why do you even want to expose `CppRef` as a separate concept?

Author: Two ergonomic reasons:
* `CppRef<T>` itself is a fairly rich type, with its own facilities for adjusting lifetimes, casts, etc. We can put all that into a trait that's implemented by every concrete C++ reference type, but it doesn't feel natural.
* It would separate `ConcreteCppType` from `ConcreteCppTypeRef`. That feels unnatural too. In pure-native Rust, we know that `T` is related to `&T`, or that `T` is related to `Arc<T>`. It doesn't seem reasonable to ask everyone to use entirely new types `RefT` or `ArcT` instead of parameterizing the smart pointer type by the pointee. The same applies in this domain.
@davidhewitt would you like to add anything?

Reviewer: > It would separate ConcreteCppType from ConcreteCppTypeRef. 

I assume ConcreteCppType wouldn't show up at all in such a case. Is it common to have APIs that have the same type both by-val and by-ref?

> CppRef<T> itself is a fairly rich type, with its own facilities for adjusting lifetimes, casts, etc. We can put all that into a trait that's implemented by every concrete C++ reference type, but it doesn't feel natural.

Can you elaborate on that? C++ references don't seem to have a ton of structure we can use here.

It might be good to link to some repository where a fully-featured `CppRef` type assuming this RFC is worked out, so that one can get an idea of how these facilities really enable useful safe abstraction. (This wouldn't be the first RFC with such an associated motivating example.) After all, as far as I can tell, the motivation entirely rests on having such a type, so working this out in full detail is crucial to justify why the existing language features do not suffice.

Author: > s it common to have APIs that have the same type both by-val and by-ref?

Yes - [C++ objects can even be kept on the Rust stack thanks to some fancy move semantics work](https://github.com/google/moveit) though it's more common to [keep them in abstractions of C++ heap pointer storage types](https://cxx.rs/binding/uniqueptr.html).

> It might be good to link to some repository where a fully-featured `CppRef` type assuming this RFC is worked out

[Here's one for C++](https://github.com/google/autocxx/blob/main/src/reference_wrapper.rs), though this does traitful workarounds for the lack of this RFC. It's already linked from the RFC. I know that this isn't the only C++ interop tool that intends to use this RFC to handle C++ references. @davidhewitt do you have a demo PyO3 equivalent?

Reviewer: Thanks for the link. However, I am confused: that `CppRef` has a safe constructor for raw pointers! So clearly it can't be used to build a safe abstraction, as that would then deref raw pointers.

You said earlier that the goal of this RFC is to make it possible to wrap C++ APIs in safe abstractions. That means making C++ methods safe to call. `CppRef` is clearly unsuited for that usecase, anyone implementing a method on `CppRef<Self>` *has to* make that method unsafe.

Author: > clearly it can't be used to build a safe abstraction, as that would then deref raw pointers.

It doesn't deref raw pointers _in Rust_. It does pass those pointers back to C++, which might do arbitrary things to them. Imagine that instead of taking a raw pointer, it took a `usize`.

You might argue that _any_ call into C++ is fundamentally unsafe because it might do arbitrary bad things. I agree! I'd argue that the fundamental unsafe-ness of C++ isn't changed by whether we're passing a pointer into it via some Rust code (which doesn't touch or dereference that pointer). So, whether C++ functions are marked as `unsafe` or not is orthogonal to this RFC. It's a [long standing debate](https://steveklabnik.com/writing/the-cxx-debate) and I'd strongly prefer not to relitigate it on this RFC. Even if one believes that all calls into C++ need to be marked `unsafe`, it's still desirable to make as-safe-as-possible wrappers on top of those C++ calls, to try to avoid C++ crashes wherever possible.

Reviewer: > It doesn't deref raw pointers in Rust. It does pass those pointers back to C++, which might do arbitrary things to them. Imagine that instead of taking a raw pointer, it took a usize.

Yes, so what?

It is the job of a safe Rust wrapper around a C++ API to know the preconditions of the C++ code, and to check them all on the Rust side. That's what being a safe wrapper *means*. `CppRef` is unsuited for that purpose. And without `CppRef` being suited for that purpose I don't see what the motivation is for this entire RFC.

The cxx debate is different. The cxx debate is about whether such a promise should be made in general by a binding tool. But it is completely accepted that a concrete C/C++/anything function can be wrapped in a safe Rust caller. Even the standard library does it when it calls `read` or `write` or any of the other C functions that it calls.

Currently the RFC claims that it aims to provide a safe wrapper around C++ API, but then fails to actually show that it is useful in achieving that goal.

Author: OK. I agree the RFC shouldn't claim it can provide a fully _safe_ wrapper around a C++ API, because (as I think we agree) a tool like bindgen can't make those promises (and this RFC is aimed at such automated tools). I will edit to ensure there are no such claims.

But "safe*r*" is important. In practice, a type like `CppRef<T>` can model C++ reference behavior far better than `&T`.

Overall I believe:

* there's going to be C++/Rust interop
* some of it is going to be achieved by automated binding generators
* C++ types will be represented as Rust types
* people are going to call C++ methods on those types.

Given those assumptions, I believe this RFC (and something like `CppRef`) is a step forward for safety. Right now, in the very widely-used cxx, `&T` is used as references to C++ types which may mutate underneath Rust at any time [0]. `CppRef<T>` is pragmatically less bad.

Let's also not forget that this is not just about C++. The Python community wants this for similar reasons - @davidhewitt might be able to provide an example which doesn't cause quite so much concern as typical C++ unsafety.

[0] = in most cases cxx manages to make the Rust type into a ZST so there's no UB... but not all.

Reviewer: It is the "automatically generated" part that I previously was missing, and that lead to us talking past each other for a bit here. Good that we reached the bottom of this!

So, I would suggest you update the motivating example in the RFC to mention all these things -- in particular the aspect about having a tool that translates a C++ reference into *some* Rust type, automatically.

I realize there's still one point I am confused though: I assume such a translation would use Rust `self` whenever there is a method with `this` in C++. But `this` has pointer type, not reference type, right? So when would such auto-generated bindings have `self: CppRef<Self>` anywhere?

> Let's also not forget that this is not just about C++. The Python community wants this for similar reasons - @davidhewitt might be able to provide an example which doesn't cause quite so much concern as typical C++ unsafety.

Sure, happy to hear about other usecases as well. :)

Author: Thanks, I'll adjust as you suggest!

Reviewer: There's still an open question. :)

"I realize there's still one point I am confused about though: I assume such a translation would use Rust self whenever there is a method with this in C++. But this has pointer type, not reference type, right? So when would such auto-generated bindings have self: CppRef\<Self> anywhere?"

I get that `type& x` in C++ becomes `x: CppRef<type>`, but where does the `self` come in? Is the argument that because `this` is non-null it is basically a reference even though it technically has pointer type?

Author: > `this` has a pointer type:

Yes, don't focus too much on the name `CppRef<T>`. In fact, one of the other bindings generation tools wanting this feature plans to call a similar thing `CppPtr<T>`, and it wouldn't surprise me if the ecosystem eventually converges around separate `CppRef<T>` and `CppPtr<T>` types. As you say, `this` is a bit more like a C++ reference in that it's supposed to be non-null (though C++ being C++, sometimes it is null, alas).

The main thing is, we can't use existing Rust types for any of this:
* Rust references have semantics which are incompatible with 
* Without arbitrary self types, we can't call methods on a Rust pointer type. And even if we could, we'd like to wrap them up in a newtype wrapper (to add functionality, check invariants wherever possible, and reduce the risk that users might get these pointers mixed up with pointers to Rust data in unsafe code)

Reviewer 7: @adetaylor @RalfJung sorry to have been slow to reply on this thread. The PyO3 use case can be summarised by a smart pointer `Py<T>` which represents data on the Python side. (For the sake of separating from `CppRef<T>` here, `Py<T>` is `unsafe` to construct, once constructed it exposes an API to call Python types.)

The type `T` is usually opaque and we don't want `Py<T>` to deref to `T` for various reasons. We would still like to be able to use `Py<T>` as a receiver for methods defined for `T`.

For example, we have `Py<PyAny>`, which has methods applicable for any Python object. We also have `Py<PyList>`, which should have the _same_ methods as `Py<PyAny>` _plus_ additional methods which are only meaningful to lists.

In current stable Rust we can model this using traits, however arbitrary self types looks like a much cleaner way to define the complete interfaces for these types `T`. At the moment associated methods like `PyList::new` are the only ones which get defined on `T`, and the rest must be defined in traits. This spilts documentation, for example.

There are various discussion threads that I can link you to if you want to read into this in more detail. The PyO3 API is going under a redesign at the moment to transition towards this `Py<T>` + traits design, so documentation will not align with what's described here.

Reviewer: Thanks for updating the RFC. :)

Reviewer 5: If I may add, in PyO3 you can have `Py<T>` where `T` is a user-defined type, which imo is an even stronger case for arbitrary self types

Reviewer 7: Yes, very true. Beyond just wanting this for ourselves, it would be very attractive for users to be able to add methods to their user-defined Python types.

    type Target = T;
}

// generated by bindings generator
struct ConcreteCppType {
    // ...
}

// all generated by bindings generator; mostly calls into C++
// In this example these are not marked "unsafe" because we do not dereference
// CppRef::ptr in Rust. However, arbitrary bad things may occur in foreign
// languages so such functions are not fully "safe" either. Safety of FFI
// is orthogonal to this RFC.
impl ConcreteCppType {
    fn some_cpp_method(self: CppRef<Self>) {}
    fn get_int_field(self: &CppRef<Self>) -> u32 {}
    fn get_more_complex_field(self: &CppRef<Self>) -> CppRef<FieldType> {}
    fn equals(self: &CppRef<Self>) -> bool {}
}

// generated by bindings generator
fn get_cpp_reference() -> CppRef<ConcreteCppType> {
    // also calls into C++
}

fn main() {
    // Rust code manipulating C++ objects via C++-semantics references
    let cpp_obj_reference: CppRef<ConcreteCppType> = get_cpp_reference();
    // cpp_obj_reference does not obey Rust reference semantics. Other
    // "references" to the same data may exist in the Rust or C++ domain.
    // But it can effectively be used as an opaque token to pass safely
    // through Rust back into C++

>     let some_value: u32 = cpp_pbj_refence.get_int_field();

Reviewer 18: ```suggestion
    let some_value: u32 = cpp_obj_reference.get_int_field();
```

    let some_field = cpp_obj_reference.get_more_complex_field();

>     cpp_obj_reference.compare_with(&get_cpp_reference());

Reviewer 10: Nit: Method is named differently above.

```suggestion
    cpp_obj_reference.equals(&get_cpp_reference());
```

For this reason, implementing `Deref::deref` is problematic for many of the likely users of this "arbitrary self types" feature.

If you're implementing a smart pointer `P<T>`, and you need to allow `impl T { fn method(self: P<T>) { ... }}`, yet you can't allow a reference `&T` to exist, any option for implementing `Deref::deref` has drawbacks:

* Specify `Deref::Target=T` and panic in `Deref::deref`. Not good.
* Specify `Deref::Target=*const T`. This is only possible if your smart pointer type contains a `*const T` which you can reference - this isn't the case for (for instance) weak pointers or types containing `NonNull`.

Therefore, the current Arbitrary Self Types v2 provides a separate `Receiver` trait, so that there's no need to provide an awkward `Deref::deref` implementation.

> In addition, this v2 proposes to block generic receivers, which are currently allowed by the v1 (unstable) arbitrary self types feature. See the [diagnostics section for reasoning](#diagnostics).

Reviewer 18: It might be worth elaborating on which parts of the receiver are allowed to be generic, and which parts aren't. Clearly, for diagnostic reasons, `fn<T: Receiver<Target = Self>>(self: T)` is problematic, but I'd expect that `Self` itself is permitted to be a generic type, as long as the smart pointer is definite; if not, that should probably be called out explicitly. Also, it might be a bit surprising to users if things like `fn<A: Allocator>(Box<Self, A>)` are disallowed by this rule, but I don't know whether that would run into the same issues as full generic receivers.

- pointers: `*const Self` and `*mut Self`

Shorthand exists for references, so that `self` with no ascription is of type `Self`, `&self` is of type `&Self` and `&mut self` is of type `&mut Self`.

All of the following self types are valid:

```rust
impl Foo {
    fn by_value(self /* self: Self */);
    fn by_ref(&self /* self: &Self */);
    fn by_ref_mut(&mut self /* self: &mut Self */);
    fn by_ptr(self: *const Self);
    fn by_mut_ptr(self: *mut Self);
    fn by_box(self: Box<Self>);
    fn by_rc(self: Rc<Self>);
    fn by_custom_ptr(self: CustomPtr<Self>);
}

struct CustomPtr<T>(*const T);

impl<T> Receiver for CustomPtr<T> {
    type Target = T;
}
```

## Recursive arbitrary receivers

Receivers are recursive and therefore allowed to be nested. If type `T` implements `Receiver<Target=U>`, and type `U` implements `Receiver<Target=Self>`, `T` is a valid receiver (and so on outward). This is the behavior for the current special-cased self types (`Pin`, `Box` etc.) so as we remove the special-casing we need to retain this property.

For example, this self type is valid:

```rust
impl MyType {
     fn by_rc_to_box(self: Rc<Box<Self>>) { ... }
}
```

> The Rust language doesn't provide a way for user code to explore this recursion, so this trait is unlikely to be useful except to the compiler. Nevertheless, we don't intend to _prevent_ use of the `Receiver` trait by user code: since the same recursive property applies to `Deref` yet it's been occasionally useful to [introduce `Deref` bounds](https://doc.rust-lang.org/std/pin/struct.Pin.html#method.new_unchecked).

Reviewer 10: Nit: I was initially confused by this paragraph. My thinking was something like "clearly, user code _can_ use the recursion, you just said that `self: Rc<Box<Self>>` is valid!"

Maybe something mentioning that they can't explore it _generically_:
```suggestion
The Rust language doesn't provide a way for user code to use this recursive property in generics, so this trait is unlikely to be useful except to the compiler. Nevertheless, we don't intend to _prevent_ use of the `Receiver` trait by user code: since the same property applies to `Deref` yet it's been occasionally useful to [introduce `Deref` bounds](https://doc.rust-lang.org/std/pin/struct.Pin.html#method.new_unchecked).
```

Author: I think what I meant was "The Rust language doesn't provide a way for user code to iterate over this recursion" - do you think that's better?

Reviewer 10: Yeah, but I think I explicitly wanted it stated in terms of "you can't add a trait bound that does this kind of recursion"

Author: Done in f51948c4169cb30dec56bcf0ecaf7a4c2d5118a9

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## `core` libs changes

The `Receiver` trait is made public (removing its `#[doc(hidden)])` attribute), exposing it under `core::ops`. It gains a `Target` associated type.

This trait marks types that can be used as receivers other than the `Self` type of an impl or trait definition.

```rust
pub trait Receiver {
    type Target: ?Sized;
}
```

A blanket implementation is provided for any type that implements `Deref`:

```rust

> impl<P: ?Sized, T: ?Sized> Receiver for P
> where
>     P: Deref<Target = T>,
> {
>     type Target = T;
> }

Reviewer 19: ```suggestion
impl<P> Receiver for P
where
    P: Deref + ?Sized,
{
    type Target = P::Target;
}
```

Reviewer 6: ```suggestion
impl<P: ?Sized> Receiver for P
where
    P: Deref,
{
    type Target = <P as Deref>::Target;
}
```
What is about this version? I find it slightly more clear.

```

(See [alternatives](#no-blanket-implementation) for discussion of the tradeoffs here.)

It is also implemented for `&T`, `&mut T`, `*const T` and `*mut T`.

## Compiler changes

The existing Rust [reference section for method calls describes the algorithm for assembling method call candidates](https://doc.rust-lang.org/reference/expressions/method-call-expr.html). This algorithm changes in one simple way: instead of dereferencing types (using the `Deref<Target=T>`) trait, we use the new `Receiver<Target=T>` trait to determine the next step.

(Note that the existing algorithm isn't quite as simple as following the chain of `Deref<Target=T>`. In particular, `&T` and `&mut T` are considered as candidates too at each step; this RFC does not change that.)

Because a blanket implementation is provided for users of the `Deref` trait and for `&T`/`&mut T`, the net behavior is similar. But this provides the opportunity for types which can't implement `Deref` to act as method receivers.

> Dereferencing a raw pointer usually needs `unsafe` (for good reason!) but in this case, no actual dereferencing occurs. This is used only to determine a list of method candidates; no memory access is performed and thus no `unsafe` is needed.

Reviewer 10: Nit: This paragraph feels a bit out of context; perhaps it needs to be framed as "the trait is not `unsafe` for this reason"?

- The current unstable arbitrary self types feature allows generic receivers. For instance,
  ```rust
  impl Foo {
    fn a<R: Deref<Target=Self>>(self: R) { }
  }
  ```
  We don't know a use-case for this. There are several cases where this can result in misleading diagnostics. (For instance, if such a method is called with an incorrect type (for example `smart_ptr.a::<&Foo>()` instead of `smart_ptr.a::<Foo>()`). We could attempt to find and fix all those cases. However, we feel that generic receiver types might risk subtle interactions with method resolutions and other parts of the language. We think it is a safer choice to generate an error on any declaration of a generic `self` type.

# Drawbacks
[drawbacks]: #drawbacks

Why should we *not* do this?

- Deref coercions can already be confusing and unexpected. Adding a new `Receiver` trait could cause similar confusion.
- Custom smart pointers are a niche use case (but they're very important for cross-language interoperability.)

## Method shadowing
[method-shadowing]: #method-shadowing

> For a smart pointer `P<T>`, a method call `p.m()` might call a method on the smart pointer type itself (`P::m`), or, if the smart pointer implements `Deref<Target=T>`, it might already call `T::m`. This already gives the possibility that `T::m` would be shadowed by `P::m`.
> 
> Current Rust standard library smart pointers are designed with this shadowing behavior in mind:
> 
> * `Box`, `Pin`, `Rc` and `Arc` already heavily use associated functions rather than methods
> * Where they use methods, it's often with the _intention_ of shadowing a method in the inner type (e.g. `Arc::clone`)
> 
> These method shadowing risks are effectively the same for `Deref` and `Receiver`. This RFC does not make things worse (it just adds additional flexibility to the `self` parameter type for `T::m`). However it does mean that the `Receiver` trait cannot be added to smart pointer types which were not designed with these concerns in mind.

Reviewer 10: Rewrite this section to have more external links. Should resolve the comment thread https://github.com/rust-lang/rfcs/pull/3519#pullrequestreview-1709560172.

```suggestion
For a smart pointer `P<T>` that implements `Deref<Target = T>`, a method call `p.m()` might call a method `P::m` on the smart pointer type itself, or it might call `T::m`. If both methods are declared, this results in an error.

Rust standard library smart pointers are designed with this shadowing behavior in mind:

* `Box`, `Pin`, `Rc` and `Arc` heavily use associated functions rather than methods.
* Where they use methods, it's often with the _intention_ of shadowing a method in the inner type (e.g. `Arc::clone`).

Furthermore, the `Deref` trait itself [documents this possible compatibility hazard](https://doc.rust-lang.org/nightly/std/ops/trait.Deref.html#when-to-implement-deref-or-derefmut), and the Rust API Guidelines has [a guideline about avoiding inherent methods on smart pointers](https://rust-lang.github.io/api-guidelines/predictability.html#smart-pointers-do-not-add-inherent-methods-c-smart-ptr).

These method shadowing risks also apply to `Receiver`. This RFC does not make things worse for types that implement `Deref`, it only adds additional flexibility to the `self` parameter type for `T::m`.
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## `core` libs changes

The `Receiver` trait is made public (removing its `#[doc(hidden)])` attribute), exposing it under `core::ops`. It gains a `Target` associated type.

This trait marks types that can be used as receivers other than the `Self` type of an impl or trait definition.

```rust
pub trait Receiver {
    type Target: ?Sized;
}
```

A blanket implementation is provided for any type that implements `Deref`:

```rust
impl<P: ?Sized, T: ?Sized> Receiver for P
where
    P: Deref<Target = T>,
{
    type Target = T;
}
```

(See [alternatives](#no-blanket-implementation) for discussion of the tradeoffs here.)

It is also implemented for `&T`, `&mut T`, `*const T` and `*mut T`.

## Compiler changes

> The existing Rust [reference section for method calls describes the algorithm for assembling method call candidates](https://doc.rust-lang.org/reference/expressions/method-call-expr.html). This algorithm changes in one simple way: instead of dereferencing types (using the `Deref<Target=T>`) trait, we use the new `Receiver<Target=T>` trait to determine the next step.
> 
> (Note that the existing algorithm isn't quite as simple as following the chain of `Deref<Target=T>`. In particular, `&T` and `&mut T` are considered as candidates too at each step; this RFC does not change that.)
> 
> Because a blanket implementation is provided for users of the `Deref` trait and for `&T`/`&mut T`, the net behavior is similar. But this provides the opportunity for types which can't implement `Deref` to act as method receivers.
> 
> Dereferencing a raw pointer usually needs `unsafe` (for good reason!) but in this case, no actual dereferencing occurs. This is used only to determine a list of method candidates; no memory access is performed and thus no `unsafe` is needed.

Reviewer 10: > > Another note: I'd like to see some more details of how method resolution and deref coercion would work with this change. Currently, it says that the algorithm would remain the same, but instead of using the Deref trait, it would use the new Receiver trait. I'm not sure this is quite right, because some method calls require deref coercions, and this new trait does not imply Deref.
> > 
> > In the following example, the `Deref` trait would still need to be involved to check if the calls to `frobnicate_ref` are allowed.
> > 
> > ```rust
> > struct Foo;
> > 
> > impl Foo{
> >     fn frobnicate_ref(&self) {}
> >     fn frobnicate_cpp_ref(self: CppRef<Self>) {}
> > }
> > 
> > fn main() {
> >     let foo_rc: Rc<Foo> = Rc::new(Foo);
> > 
> >     // this compiles fine, and desugars to `Foo::frobnicate_ref(&*foo_rc)`
> >     foo_rc.frobnicate_ref();
> > 
> >     let foo_cpp_ref: CppRef<Foo> = CppRef::new(Foo);
> > 
> >     // should not compile because it would desugar to `Foo::frobnicate_ref(&*foo_cpp_ref)`
> >     // and you can't deref a CppRef
> >     foo_cpp_ref.frobnicate_ref();
> > 
> >     // should compile, because we're not dereffing the CppRef
> >     // desugars to `Foo::frobnicate_cpp_ref(foo_cpp_ref)`
> >     foo_cpp_ref.frobnicate_cpp_ref();
> > }
> > ```
> 
> Yeah, I caught that too.
> 
> I think the RFC is mixing two sets of semantics at times. As far as I understand, [the reference](https://doc.rust-lang.org/reference/expressions/method-call-expr.html) as it's currently written doesn't explicitly address Box/Arc/etc as receivers, but it does implicitly address them in this paragraph:
> 
> > Then, for each candidate type T, search for a [visible](https://doc.rust-lang.org/reference/visibility-and-privacy.html) method with a receiver of that type in the following places:
> > 
> > * T's inherent methods (methods implemented directly on T).
> > * Any of the methods provided by a [visible](https://doc.rust-lang.org/reference/visibility-and-privacy.html) trait implemented by T. If T is a type parameter, methods provided by trait bounds on T are looked up first. Then all remaining methods in scope are looked up.
> 
> That first bullet point is the part that must be amended to clarify that "T's inherent methods" need to take the Receiver trait into account; the paragraph above about dereferencing chains must be kept untouched.

Huh, yeah, took me like two hours to properly consider this, but I think you've really got the gist of it here! Essentially, `Receiver` should _restrict_ which inherent implementations one can make, but doesn't actually by itself affect method resolution!

How about we change this section to instead be something like:

````suggestion
For the uninitiated, the following impl briefly describes the existing terminology:

```rust
impl ImplementingType {
    fn foo(self: ReceiverType);
}
```

---

Instead of being hard-coded in the compiler, the `Receiver` trait now limits which types are allowed as method receivers. That is, [the allowed types of the `self` parameter in methods](https://doc.rust-lang.org/reference/items/associated-items.html#methods) are now limited to those generated by the following grammar:

```text
P = any type that implements Receiver<Target = S>
S = Self | P
```

The existing Rust [reference section for method calls describes the algorithm for assembling method call candidates](https://doc.rust-lang.org/reference/expressions/method-call-expr.html). This algorithm changes slightly: instead of only searching `T`'s own inherent methods, any method that has a receiver with the given type is considered.

Note that what the implementing type for a method is (whether that's `T`, `&T`, `CppRef<T>` or `&CppRef<T>`) doesn't affect method resolution, only the receiver type matters in that regard.
````

(The rest of the RFC will likely need changes to accommodate for this, too).

Reviewer 10: Extending the given frobnicate example with more comments, assuming the above change:
```rust
// `foo_rc: Rc<Foo>`, so the possible method candidates are:
// - `Rc<Foo>`
// - `&Rc<Foo>`
// - `&mut Rc<Foo>`
// - `Foo` (by dereferencing)
// - `&Foo`
// - `&mut Foo`
//
// There exists a visible method named `frobnicate_ref` with receiver type `&Foo`, so that is selected.
foo_rc.frobnicate_ref();

// `foo_cpp_ref: CppRef<Foo>`, so the possible method candidates are:
// - `CppRef<Foo>`
// - `&CppRef<Foo>`
// - `&mut CppRef<Foo>`
//
// There are no methods named `frobnicate_ref` with any of the above receiver types, so this is an error.
foo_cpp_ref.frobnicate_ref();

// `foo_cpp_ref` is the same as above, so the method candidates are the same.
//
// In this case though, there exists a visible method named `frobnicate_cpp_ref` with receiver type `&CppRef<Foo>`, so that is selected.
// Note that whether the method is implemented on `Foo`, `CppRef<Foo>` or `&CppRef<Foo>` doesn't matter.
foo_cpp_ref.frobnicate_cpp_ref();
```

This also explains what happens if we were to add the impl below: There would suddenly be two inherent methods with the same name that has receiver type `&CppRef<Foo>`, and that's an error.

```rust
impl<T> CppRef<T> {
    fn frobnicate_cpp_ref(&self) {}
}
```

Author: I'm glad to see a more precise Reference section in this RFC, I've felt uncomfortable about it for a while, and I'd like to commit the above suggested change.

The code which is changing for this RFC is indeed in the section of code which assembles method call _candidates_ which I think is another way of saying:
> Essentially, Receiver should restrict which inherent implementations one can make, but doesn't actually by itself affect method resolution!
so I think the understanding here is correct.

@madsmtm please would you have a look at `wfcheck.rs` in the [hypothetical semi-correct implementation of this RFC](https://github.com/rust-lang/rust/compare/master...adetaylor:rust:receiver_trait_with_target_20231117) to see if it matches your renewed understanding? If so I'll go ahead and commit this change, and we can work on the rest of the wording over time.

Reviewer 10: I'm... not really well-enough versed in `rustc` internals to say for sure, though I think it looks correct! (my mental reasoning: `wfcheck.rs` is (amongst a lot of other stuff) checking whether the implementation is allowed or not based on `Receiver`, allowing e.g. `self: CppRef<Self>` but disallowing e.g. `self: u8`).

> The code which is changing for this RFC is indeed in the section of code which assembles method call _candidates_ which I think is another way of saying *snip*

Yup, the receiver candidate selection stays exactly the same, only the inherent method candidate selection changes.

Author: OK I had a play around. Our current working branch is somewhat naïve in this respect and we definitely need to be a bit more sophisticated here. I haven't got it right yet. [Here's a comment explaining the remaining stuff that I know we need to figure out.](https://github.com/rust-lang/rust/commit/32c1585788d51d2e6a4fa7e1c0c30cf4461c6477)

I haven't yet convinced myself whether your revised description above is accurate, so I'm not going to commit it yet. I feel that it definitely needs to describe the search algorithm a bit more programmatically. Right now it says "any method that has a receiver of the given type is considered" - but which `impl` blocks does it search to find them?

It's definitely the case that the statement "we just change the algorithm in one simple way" alas isn't quite true though, so we'll definitely make a change here.

Reviewer 10: > but which `impl` blocks does it search to find them?

I think I was kinda thinking... All of them? Though clearly that's ill-defined for cross-crate compilation, so we'll need something more precise here - I am unsure what that should be though, will give it some more thought myself!

Reviewer 10: I also missed a nuance above: In [the original text for allowed `self` types](https://doc.rust-lang.org/reference/items/associated-items.html#methods), `Pin<P>` is actually only allowed for pointer types, e.g. `self: Pin<Self>` is disallowed; unsure if that's something we'd want to preserve?

I'm still fairly convinced that modifying the candidate type search is not the way to go, effectively that part is just saying "which types can I feasibly reach by inserting derefs or a borrow", which is important for when we get to actually desugaring the method call.

Author: (Update: this is the bit I'm trying to work on in code right now, so I can validate if the above idea actually works or is nonsense. I'm travelling quite a bit for work at the moment so not getting a lot of time, but hopefully sometime in the next few days we'll find out)

## No blanket implementation for `Deref`
[no-blanket-implementation]: #no-blanket-implementation

The other major approach previously discussed is to have a `Receiver` trait, as proposed in this RFC, but without a blanket implementation for `T: Deref`. Blanket implementations are unusual for core Rust traits, but the authors of this RFC believe it's necessary in this case.

Specifically, this RFC proposes that the existing method search algorithm is modified to search the `Receiver` chain _instead of_ the `Deref` chain.

It's therefore a major compatibility break if existing `Deref` implementors cease to be usable as `self` parameters. Just in the standard library, we'd have to add `Receiver` implementations for `Cow`, `Ref`, `ManuallyDrop` and possibly many other existing implementors of `Deref`: third party libraries would have to do the same. Without that, method calls on these types would not be possible:

```rust
fn main() {
    let ref_cell = RefCell::new(/* something cloneable */);
    ref_cell.borrow().clone(); // no longer possible if:
        // 1) we cease to explore Deref in identifying method candidates
        // 2) Ref doesn't implement Receiver.
}
```

> This doesn't just break people previously using the unstable Rust `arbitrary_self_type` feature; it breaks stable Rust usages as well. Obviously this is not acceptable, so we believe the blanket implementation is necessary.

Reviewer 16: > In this case though the authors of this RFC believe it necessary, because the RFC proposes that the existing method search algorithm is modified to search the `Receiver` chain _instead of_ the `Deref` chain, so it would be a major compatibility break if existing `Deref` implementors would suddenly be required to also implement `Receiver` as well to get the current behaviour.

But... it still *doesn't* search the receiver chain, right? Like you said, resolution first looks through the deref chain, and then for each method, looks if the receiver type is allowed, based on the receiver trait?

> This doesn't just break people previously using the unstable Rust `arbitrary_self_type` feature; it breaks stable Rust usages as well. Obviously this is not acceptable, so we believe the blanket implementation is necessary.

Reviewer 10: Yeah, I wrote this before I wrote https://github.com/rust-lang/rfcs/pull/3519#discussion_r1400808795, this will need some modifications after that.

> This doesn't just break people previously using the unstable Rust `arbitrary_self_type` feature; it breaks stable Rust usages as well. Obviously this is not acceptable, so we believe the blanket implementation is necessary.

Reviewer 14: @madsmtm In theory, it *would* be possible to introduce a mechanism that'd allow introducing new supertraits for a trait, without breaking backwards compatibility.

Imagine if we had something like "inherent supertraits", such that we could say the `Target` type in `Deref` is actually part of `Receiver`, and *if* you write just an `impl Deref for T` that includes `Target` then that implicitly defines `Receiver`, while you also have the option of writing an explicit `impl Receiver for T` in which case you can't also define `Target` when defining `impl Deref for T`.

Effectively, this would give the option of "splitting out" some methods or associated types of a trait into a supertrait, compatibly, such that anyone who acts like they're *not* split out is still correct, and anyone who acts like they *are* split out is also correct.

This wouldn't be worth it for `Receiver`/`Deref` alone, but it seems like a potentially useful general feature. The question is then whether we think that would be *better* than the blanket impl in any way. Is there anything we could do given the supertrait approach (`Deref: Receiver`) that we *couldn't* do with the blanket impl?

> This doesn't just break people previously using the unstable Rust `arbitrary_self_type` feature; it breaks stable Rust usages as well. Obviously this is not acceptable, so we believe the blanket implementation is necessary.

Reviewer 10: > This wouldn't be worth it for `Receiver`/`Deref` alone, but it seems like a potentially useful general feature. The question is then whether we think that would be better than the blanket impl in any way. Is there anything we could do given the supertrait approach (`Deref: Receiver`) that we couldn't do with the blanket impl?

I don't think the supertrait approach gives us anything other than being more in line with how things currently work.

(E.g. to implement `PartialOrd`, you have to implement `PartialEq` first, instead of e.g. `PartialEq` having a blanket impl for `T: PartialOrd`).

Dereferencing a raw pointer usually needs `unsafe` (for good reason!) but in this case, no actual dereferencing occurs. This is used only to determine a list of method candidates; no memory access is performed and thus no `unsafe` is needed.

## Object safety

Receivers are object safe if they implement the (unstable) `core::ops::DispatchFromDyn` trait.

As not all receivers might want to permit object safety or are unable to support it, object safety should remain being encoded in a different trait than the here proposed `Receiver` trait, likely `DispatchFromDyn`.

This RFC does not propose any changes to `DispatchFromDyn`. Since `DispatchFromDyn` is unstable at the moment, object-safe receivers might be delayed until `DispatchFromDyn` is stabilized. `Receiver` is not blocked on further `DispatchFromDyn` work, since non-object-safe receivers already cover a big chunk of the use-cases.

## Lifetime elision

Arbitrary `self` parameters may involve lifetimes.

> This RFC currently proposes no changes to the standard lifetime elision rules. There are no known cases where ambiguity results. However, the opposite problem does apply: sometimes explicit lifetimes are required for methods with an arbitrary self type, even if they're not required for an equivalent free function. 

Reviewer: This is still an ill-formed statement. The existing rules can't even apply to your new examples. You are extending the domain of legal function declarations that elision has to work on, you have to say how it behaves on that extended domain!

Author: > You are extending the domain of legal function declarations that elision has to work on

My assumption (which may be wrong) is that `self: Something` behaves identically to `any_other_param: Something`, which is why I don't see it as extending the domain of legal functions. It's just a parameter rename.

But, I'm going to proceed as if that assumption is wrong (some of the results in my example below suggest it is wrong) so I'll get my head around the existing lifetime elision code to explain what's up. This may take some time.

Reviewer: That assumption is wrong. `fn foo(x: &T, y: &U) -> &Z` is invalid, but `fn foo(&self, y: &U) -> &Z` is allowed. The `self` parameter is special, if present then an elided lifetime in the return type will default to the lifetime of `self` even if there are other lifetimes in the signature.

Please have a look at https://doc.rust-lang.org/beta/reference/lifetime-elision.html :)

Author: Yep, thanks for working through this with me.

The current method-specific rule is:
> If the receiver has type `&Self` or `&mut Self`, then the lifetime of that reference to Self is assigned to all elided output lifetime parameters.

I had forgotten about that rule, but I think it's probably wise if it doesn't apply in this case. i.e. I propose that this rule continues to strictly apply _only_ to `&Self` and `&mut Self`, then no lifetime elision applies to `self: AnythingElse<T>` nor `self: &AnythingElse<T>` nor `self: &mut AnythingElse<T>`? That is, in the case of an arbitrary-self-type receiver, no elision occurs.

This feels like a case (rather like blocking generic `Receiver`s for now) where we can simplify the problem space for now, and later relax these rules if use-cases present themselves?

WDYT?

Reviewer: I think that documentation is outdated, because this compiles:
```rust
use std::rc::Rc;

struct S;

impl S {
    fn test(self: Rc<&Self>) -> &i32 { todo!() }
}
```

Author: Ah OK. I'll probably aim to fix the current documentation (and thus come to understand the current behavior) before returning to this comment thread on this RFC. I'll leave this comment thread open but resolve the other two threads related to lifetime elision, since they're all related.

Author: Still digging but leaving notes on the current rules (in _stable_ Rust) because I'm finding them surprising.

With `struct Concrete(u32);` and all the following methods in an `impl Concrete` block.

Does not compile:
```rust
fn n(self: &Box<Self>, _other: &u32) -> &u32 {
    &self.0
}
```
This gives an error saying that we're returning data with the lifetime of `_other`; that is, the lifetime elision rules have not spotted the lifetime attached to `self`, and assume therefore that the output lifetime must belong to the other parameter. Adding an explicit lifetime solves this. Arguably this is a bug? Though I can't see when it would have actual bad consequences.

Also does not compile, and gives a curious message:
```rust
fn n1(self: &Box<Self>) -> &u32 {
    &self.0
}
```
Says "this function's return type contains a borrowed value, but the signature does not say which `self` it is borrowed from". Adding an explicit lifetime solves this.

That diagnostic, at least, feels non-optimal.

Does compile:

```rust
fn m(self: Box<&Self>, _other: &u32) -> &u32 {
    &self.0
}
fn g(self: Box<&Self>, _other: &u32) -> &u32 {
    &self.0
}
fn g1<'a, 'b>(self: &'a Box<&'b Self>, _other: &u32) -> &'a u32 {
    &self.0
}
fn g2<'a, 'b>(self: &'a Box<&'b Self>, _other: &u32) -> &'b u32 {
  &self.0
}
```

`g` seems to be ambiguous, although in practice Rust seems able to prove that lifetimes a and b are equivalent (hence both `g1` and `g2` compiling). It makes a difference to callers. `g` has the same semantics as `g2`.

Reviewer 5: `'a` and `'b` are not equivalent. Two things are happening:
- the type of `self: &'a Box<&'b Self>` adds an implicit `'b: 'a` bound, which means that the inner value will stay live as long as `'a` is. This makes returning `&'a u32` correct.
- `self.0` copies a `&'b Self` out of the box, after which we don't need the box anymore. This allows us to return `&'b u32`.

It's not the same however: `g2` is more general than `g1` since `'a` could be much shorter than `'b`.

It seems lifetime elision is doing something like "take the lifetime of the innermost reference", which does infer the most general lifetime. It feels like a good general rule? It likely won't generalize to more general `Receiver`s though.

Reviewer: > This gives an error saying that we're returning data with the lifetime of _other; that is, the lifetime elision rules have not spotted the lifetime attached to self, and assume therefore that the output lifetime must belong to the other parameter. Adding an explicit lifetime solves this. Arguably this is a bug? Though I can't see when it would have actual bad consequences.

That's a bug indeed, good catch!

Author: I have rewritten this section to say pretty much:

* There are already confusing/wrong lifetime situations in stable Rust today
* Please discuss them on the [bug report](https://github.com/rust-lang/rust/issues/117715) instead of here (though I give a preview of what we're currently intending to make the rules on that bug)
* We don't intend to change anything here in the RFC

So I'm going to resolve this thread for now. Thanks for all the discussion and guiding me towards the current faulty behavior.

```rust
use std::ops::Receiver;

struct SmartPtrByValue<T: ?Sized>(T);

impl<T: ?Sized> Receiver for SmartPtrByValue<T> {
    type Target = T;
}

struct SmartPtrByRef<'a, T: ?Sized>(&'a T);

impl<'a, T: ?Sized> Receiver for SmartPtrByRef<'a, T> {
    type Target = T;
}

struct Concrete(u32);

impl Concrete {

>     // fn a(self: &SmartPtrByValue<Self>) -> &u32 { &self.0.0 } // does not compile

Reviewer: Why should this not compile? This RFC is supposed to be the normative document saying what the rules are and why. So you should give motivation when there are strange cases like this.

Author: Resolving this comment thread in favor of [this one about many of the same issues](https://github.com/rust-lang/rfcs/pull/3519#discussion_r1381792163) - in short all this lifetime elision section needs redoing.

    fn b<'a>(self: &'a SmartPtrByValue<Self>) -> &'a u32 { &self.0.0 }
    // fn c(self: &SmartPtrB<Self>) -> &u32 {} // does not compile
    fn d<'a, 'b>(self: &'a SmartPtrByRef<'b, Self>) -> &'a u32 { &self.0.0 }
    fn e<'a>(self: &'a SmartPtrByRef<Self>) -> &'a u32 { &self.0.0 }
    fn f<'a>(self: &'_ SmartPtrByRef<'a, Self>) -> &'a u32 { &self.0.0 }
    fn g<'a, 'b>(self: &'a SmartPtrByRef<'b, Self>) -> &'b u32 { &self.0.0 }
    fn h<'a>(self: SmartPtrByRef<'a, Self>) -> &'a u32 { &self.0.0 }

>     // fn i(self: SmartPtrByRef<Self>) -> &u32 { &self.0.0 } // does not compile

Reviewer: This has 0 examples of eliding the lifetime of the returned reference that actually build. So clearly there's something missing, unless you want to imply that this is never supported (but even then you'd have to say that).

Author: Resolving this comment thread in favor of [this one about many of the same issues](https://github.com/rust-lang/rfcs/pull/3519#discussion_r1381792163) - in short all this lifetime elision section needs redoing.

Specifically, that page also states:

> If this results in multiple possible candidates, then it is an error, and the receiver must be converted to an appropriate receiver type to make the method call.

This changes. For smart pointer types which implement `Receiver` (such as `NonNull<T>`) the future addition of any method would become an incompatible change, because it would run the risk of this ambiguity if there were a method of the same name within `T`. So, if there are multiple candidates, and if one of those candidates is in a more "nested" level of receiver than the others (that is, further along the chain of `Receiver`), we will choose that candidate and warn instead of producing a fatal error.

Similarly,

> Note: the lookup is done for each type in order, which can occasionally lead to surprising results.

This changes too, for the same reason. We check for matching candidates for `T`, `&T` and `&mut T`, and again, if there's a candidate on an "inner" type (that, is, further along the chain of `Receiver`) we will choose that type in preference to less nested types and emit a warning.

> (The current reference doesn't describe it, but the current algorithm also searches for method receivers of type `*const Self` and handles them explicitly in case the receiver type was `*mut Self`. We do not check for cases where a new `self: *mut Self` method on an outer type might shadow an existing `self: *const SomePtr<Self>` method on an inner type. Although this is a theoretical risk, such compatibility breaks should be easy to avoid because `self: *mut Self` are rare. It's not readily possible to apply the same de-shadowing approach to these, because we already intentionally shadow `*const::cast` with `*mut::cast`.)

Reviewer 5: Hm yes, the question becomes: can we make the simple version of arbitrary self types forward-compatible with potential future changes to method resolution we'd need for `Weak` etc?

I think yes: let's make it a hard error to call a method if there's ambiguity. That's already what happens with the current implementation of `arbitrary_self_types` on nightly. When we figure out `Weak` etc we can relax this hard error into a warning. A little bit of care might be necessary for mixes of `Deref` and `Receiver` stacks but I'm not super worried.

```rust
fn main() {
    let ref_cell = RefCell::new(/* something cloneable */);
    ref_cell.borrow().clone(); // no longer possible if:
        // 1) we cease to explore Deref in identifying method candidates
        // 2) Ref doesn't implement Receiver.
}
```

This doesn't just break people previously using the unstable Rust `arbitrary_self_type` feature; it breaks stable Rust usages as well. Obviously this is not acceptable, so we believe the blanket implementation is necessary.

In any case, we think a blanket implementation is desirable:

* It prevents `Deref` and `Receiver` having different `Target`s. That could possible lead to confusion if it prompted the compiler to explore different chains for these two different purposes.
* If smart pointer type `P<T>` is in a crate, users of `P` to create `P<MyConcreteType>` will be able to use it as a `self` type for `MyConcreteType` without waiting for a new release of the `P` crate.

> We found that [some crates use `Deref` to express an is-a not a has-a relationship](ttps://gist.github.com/davidhewitt/d0ed031fb05f6db98ee249ae089b268e) and so, ideally, might have preferred the option of setting up `Deref` and `self` candidacy separately. But, on discussion, we concluded that traits would be a better way to model those relationships.

Reviewer 20: ```suggestion
We found that [some crates use `Deref` to express an is-a not a has-a relationship](https://gist.github.com/davidhewitt/d0ed031fb05f6db98ee249ae089b268e) and so, ideally, might have preferred the option of setting up `Deref` and `self` candidacy separately. But, on discussion, we concluded that traits would be a better way to model those relationships.
```

Link was not appearing at all in github.

    // modification, so some_foreign_language_reference can't be a &T
    // But we still want to be able to do this
    some_foreign_language_reference.SomeForeignLanguageMethod(); // executes in the foreign language. Data is not
        // dereferenced at all in Rust.
}
```

Even if the reader takes the view that all calls into foreign languages are intrinsically unsafe and must be marked as such, hopefully the reader would support building abstractions using the Rust type system to minimize the practical risk of undefined behavior. That's what this RFC aims to enable.

# Prior art
[prior-art]: #prior-art

A previous PR based on the `Deref` alternative has been proposed before https://github.com/rust-lang/rfcs/pull/2362 and was postponed with the expectation that the lang team would [get back to `arbitrary_self_types` eventually](https://github.com/rust-lang/rfcs/pull/2362#issuecomment-527306157).

# Future work

As [discussed above](#pick-shadowed-methods-instead-of-erroring) we anticipate a future version which will relax some errors into warnings, and thus allow us to add support for raw pointers, `Weak` and `NonNull` as self types.

> Thereafter, we could consider implementing `Receiver` for other types, e.g. [`std::cell`](https://doc.rust-lang.org/std/cell/index.html) types, [`std::sync`](https://doc.rust-lang.org/std/sync/index.html) types, [`std::cmp::Reverse`](https://doc.rust-lang.org/std/cmp/struct.Reverse.html), [`std::num::Wrapping`](https://doc.rust-lang.org/nightly/std/num/struct.Wrapping.html), [`std::mem::MaybeUninit`](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html), [`std::task::Poll`](https://doc.rust-lang.org/nightly/std/task/enum.Poll.html), and so on - possibly even for arrays, `Vec`, `BTreeSet` etc.

Reviewer 20: Except `Vec` would not work, since it already implements `Deref`.

Author: Good point - fixed. Thanks.

The `Receiver` trait is simple and only requires specifying the `Target` type:

```rust
trait Receiver {
    type Target: ?Sized;
}
```

The `Receiver` trait is already implemented for many standard library types:
- smart pointers in the standard library: `Rc<Self>`, `Arc<Self>`, `Box<Self>`, and `Pin<SomeSmartPtr<Self>>` (and in fact, any type which implements `Deref`)
- references: `&Self` and `&mut Self`
- pointers: `*const Self` and `*mut Self`

Shorthand exists for references, so that `self` with no ascription is of type `Self`, `&self` is of type `&Self` and `&mut self` is of type `&mut Self`.

All of the following self types are valid:

```rust
impl Foo {
    fn by_value(self /* self: Self */);
    fn by_ref(&self /* self: &Self */);
    fn by_ref_mut(&mut self /* self: &mut Self */);
    fn by_ptr(self: *const Self);
    fn by_mut_ptr(self: *mut Self);
    fn by_box(self: Box<Self>);
    fn by_rc(self: Rc<Self>);
    fn by_custom_ptr(self: CustomPtr<Self>);
}

struct CustomPtr<T>(*const T);

impl<T> Receiver for CustomPtr<T> {
    type Target = T;
}
```

## Recursive arbitrary receivers

> Receivers are recursive and therefore allowed to be nested. If type `T` implements `Receiver<Target=U>`, and type `U` implements `Receiver<Target=Self>`, `T` is a valid receiver (and so on outward). This is the behavior for the current special-cased self types (`Pin`, `Box` etc.) so as we remove the special-casing we need to retain this property.

Reviewer 21: I found the last sentence here hard to read, I initially read it as:
- so as we remove, the special casing we need to retain this property _(missing words?)_.
- so ~~as~~ we remove the special casing we need to retain this property.

The third time through, I realised the phrasing was meant to be:
```suggestion
Receivers are recursive and therefore allowed to be nested. If type `T` implements `Receiver<Target=U>`, and type `U` implements `Receiver<Target=Self>`, `T` is a valid receiver (and so on outward). This is the behavior for the current special-cased self types (`Pin`, `Box` etc.), so as we remove the special-casing, we need to retain this property.
```

Another alternative is to split the long sentence:
```suggestion
Receivers are recursive and therefore allowed to be nested. If type `T` implements `Receiver<Target=U>`, and type `U` implements `Receiver<Target=Self>`, `T` is a valid receiver (and so on outward). This is the behavior for the current special-cased self types (`Pin`, `Box` etc.). So as we remove the special-casing, we need to retain this property.
```

This principle has been followed for the types in Rust's standard library which implement `Receiver`; for instance, `Box` and `Rc`. Mostly they offer associated functions rather than methods.

In the future there might be a deshadowing algorithm that can relax this rule - see the [method shadowing section below](#method-shadowing) for discussion.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## `core` libs changes

The `Receiver` trait is made public (removing its `#[doc(hidden)])` attribute), exposing it under `core::ops`. It gains a `Target` associated type.

This trait marks types that can be used as receivers other than the `Self` type of an impl or trait definition.

```rust
pub trait Receiver {
    type Target: ?Sized;
}
```

A blanket implementation is provided for any type that implements `Deref`:

```rust
impl<P: ?Sized> Receiver for P
where
    P: Deref,
{
    type Target = <P as Deref>::Target;
}
```

> (See [alternatives](#no-blanket-implementation) for discussion of the tradeoffs here.)

Reviewer 22: ```suggestion
(See [alternatives](#no-blanket-implementation-for-deref) for discussion of the tradeoffs here.)
```

> It is also implemented for `&T` and `&mut T`.

Reviewer 21: Would it help to also mention pointers `*const T` and `*mut T` here, for consistency with the Guide-level section?

Author: Thanks for the spot here - the mention of supporting `Receiver` on pointers was hanging around from earlier drafts, I've removed it.

## Motivation for the arbitrary self types feature overall

One use-case is cross-language interop (JavaScript, Python, C++), where other languages' references can’t guarantee the aliasing and exclusivity semantics required of a Rust reference. For example, the C++ `this` pointer can't be practically or safely represented as a Rust reference because C++ may retain other pointers to the data and it might mutate at any time. Yet, calling C++ methods intrinsically requires a `this` reference. With "arbitrary self types", smart pointer types can be created which obey foreign-language semantics and can be used in safe Rust code:

```rust
#[repr(transparent)]
#[derive(Clone)]
/// A C++ reference. Obeys C++ reference semantics, not Rust reference semantics.

> /// There is no exclusivity; the underlying data may mutate, etc.

Reviewer: I assume this *does* carry an invariant that the pointer points to dereferenceable memory though and holds a valid instance of the corresponding C++ type? Would be good to document explicitly what is guaranteed about the fields of this type.

Author: Maybe the semantics are just as you describe, but the point of this RFC is that _Rust doesn't care_. People should be free to implement whatever semantics they want for foreign language smart pointers.

Rust only starts to care if we dereference the data in Rust. Otherwise, the underlying pointer is just an opaque token which ephemerally passes through the Rust language, and the other language can use it however it wants to.

Reviewer: This is an example though. Everyone writing such a type should document their particular invariant, so if the example doesn't do that then it does not follow best practices.

Author: I've added a sentence explaining that (for the purposes of this RFC) the invariants are not fully documented.

Actually documenting them would distract from this RFC, because pointer and reference semantics in C++ are essentially unpredictable in a sufficiently complex codebase. There's a reasonable chance that this pointer could point to the null page, or to a deallocated object, or just to something entirely wrong. That's the inevitability of C++, and in a sufficiently complex interop scenario, there's every chance that those problems end up applying to some percentage of the `CppRef::ptr`s.

The "pointer" inside here really should be thought of as an opaque token. It might not even be a pointer. [Here's an example where it would be some index into a lookup table](https://github.com/rust-lang/rust/issues/103516#issuecomment-1790856267).

I think what you're hinting at here is a deeper concern: unless we _can_ document reasonable invariants for `CppRef::ptr`, then we should not allow that pointer to exist in Rust, because neither the compiler (in safe code) nor the human (in unsafe code) can guarantee those invariants are met. I can see that point of view. But this is to say that we should not have interop with other languages in Rust, because they intrinsically have weaker guarantees than Rust. So long as we accept that interop with other languages is a necessity, then we should allow the Rust user to build safer abstractions to reduce the practical risk of the unsafety. That's what this is all about.

Reviewer: I don't think actually documenting them would distract from the RFC at all. An important part of an RFC is to show how the RFC would be used on real code, and for using this RFC in real code, having such a safety invariant is absolutely crucial! The entire RFC rests in having raw pointers in private fields that we know extra things about due to an invariant, and relying on that invariant to provide a safe API.

Author: Note to other readers: in another comment thread we discovered that this RFC wasn't sufficiently clear that we're mostly aiming at _auto generated bindings_. Over the coming days I'll update the RFC to make that clearer, which should also end up resolving this comment thread by more accurately describing the invariants which are, and are not, realistically possible for such auto-generated bindings. I'll resolve this comment thread too once I've made that update.

Author: I hope this is addressed by 69b1edfc07a8926e7a60638a7a9cb336f2cb86ef. I'm going to mark this conversation thread as resolved, but we can continue to discuss on [the related comment thread](https://github.com/rust-lang/rfcs/pull/3519#discussion_r1380344160) if it's not sorted out.

## Motivation for the v2 changes

Unstable Rust contains an implementation of arbitrary self types based around the `Deref<Target=T>` trait. Naturally, that trait also provides a means to create a `&T`. Example:

```rust
#[feature(arbitrary_self_types)]

struct SmartPtr<T>(*const T);

impl<T: ?Sized> Deref for SmartPtr<T> {
    type Target = T;
    fn deref(&self) -> &Self::Target {
        // never called, but smart pointers need to implement this method
        // sometimes it's just not safe to create a reference to self.0
    }
}

struct ConcreteType;

impl ConcreteType {
    fn some_method(self: SmartPtr<ConcreteType>) {

    }
}

fn main() {
    let concrete: SmartPtr<ConcreteType> = ...;
    concrete.some_method();
}
```

However, if it's OK to create a reference `&T`, you _probably_ don't need this feature. You can simply use `&self` as your receiver type:

```rust
impl ConcreteType {
    fn some_method(&self) {

    }
}
```

> This feature is mostly aimed at smart pointer types `P<T>` where it's not safe to create a reference `&T`. As noted above, that's most commonly because of semantic differences to pointers in other languages, but it might be because references have special meaning or behavior in some pure Rust domain. Either way, it's not OK to create a Rust reference `&T` or `&mut T`, yet we may want to allow methods to be called on some reference-like thing.

Reviewer: I wouldn't agree with the "mostly" here. The biggest motivation for this feature was to call methods on `Pin<&mut Self>` -- methods like `Future::poll`; it is no coincidence that this was stabilized together with the entire `async`/`Future` push. `Rc<Self>` was also important.

Supporting unsafe raw pointers is a nice addition, but arbitrary self receivers is very useful even in its current form.

Author: OK. Thanks. I'll think about how to reword this.

Author: OK, I adjusted this in e269c006. I'm going to resolve this conversation so I can keep track of what still needs to be addressed, but please do reopen it if you don't think it accurately reflects the situation.
